<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
    <meta charset="utf-8">
    <meta name="generator" content="quarto-1.4.551">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

    <meta name="author" content="Kitty B Murphy">
    <meta name="author" content="Brian M Schilder">
    <meta name="author" content="Nathan G Skene">
    <meta name="dcterms.date" content="2024-05-22">
    <meta name="keywords" content="Ontology, Human Phenotype Ontology, Large Language Model, GPT-4, Artificial intelligence, Generative AI, Rare disease, Automation, Medicine">

    <title>Harnessing generative AI to annotate the severity of all phenotypic abnormalities within the Human Phenotype Ontology</title>
    <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      div.columns{display: flex; gap: min(4vw, 1.5em);}
      div.column{flex: auto; overflow-x: auto;}
      div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
      ul.task-list{list-style: none;}
      ul.task-list li input[type="checkbox"] {
        width: 0.8em;
        margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
        vertical-align: middle;
      }
      /* CSS for syntax highlighting */
      pre > code.sourceCode { white-space: pre; position: relative; }
      pre > code.sourceCode > span { line-height: 1.25; }
      pre > code.sourceCode > span:empty { height: 1.2em; }
      .sourceCode { overflow: visible; }
      code.sourceCode > span { color: inherit; text-decoration: inherit; }
      div.sourceCode { margin: 1em 0; }
      pre.sourceCode { margin: 0; }
      @media screen {
      div.sourceCode { overflow: auto; }
      }
      @media print {
      pre > code.sourceCode { white-space: pre-wrap; }
      pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
      }
      pre.numberSource code
        { counter-reset: source-line 0; }
      pre.numberSource code > span
        { position: relative; left: -4em; counter-increment: source-line; }
      pre.numberSource code > span > a:first-child::before
        { content: counter(source-line);
          position: relative; left: -1em; text-align: right; vertical-align: baseline;
          border: none; display: inline-block;
          -webkit-touch-callout: none; -webkit-user-select: none;
          -khtml-user-select: none; -moz-user-select: none;
          -ms-user-select: none; user-select: none;
          padding: 0 4px; width: 4em;
        }
      pre.numberSource { margin-left: 3em;  padding-left: 4px; }
      div.sourceCode
        {   }
      @media screen {
      pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
      }
      /* CSS for citations */
      div.csl-bib-body { }
      div.csl-entry {
        clear: both;
        margin-bottom: 0em;
      }
      .hanging-indent div.csl-entry {
        margin-left:2em;
        text-indent:-2em;
      }
      div.csl-left-margin {
        min-width:2em;
        float:left;
      }
      div.csl-right-inline {
        margin-left:2em;
        padding-left:1em;
      }
      div.csl-indent {
        margin-left: 2em;
      }    </style>

    <style>
      body.hypothesis-enabled #quarto-embed-header {
        padding-right: 36px;
      }

      #quarto-embed-header {
        height: 3em;
        width: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        border-bottom: solid 1px;
      }

      #quarto-embed-header h6 {
        font-size: 1.1em;
        padding-top: 0.6em;
        margin-left: 1em;
        margin-right: 1em;
        font-weight: 400;
      }

      #quarto-embed-header a.quarto-back-link,
      #quarto-embed-header a.quarto-download-embed {
        font-size: 0.8em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 1em;
        margin-right: 1em;
      }

      .quarto-back-container {
        padding-left: 0.5em;
        display: flex;
      }

      .headroom {
          will-change: transform;
          transition: transform 200ms linear;
      }

      .headroom--pinned {
          transform: translateY(0%);
      }

      .headroom--unpinned {
          transform: translateY(-100%);
      }      
    </style>

    <script>
    window.document.addEventListener("DOMContentLoaded", function () {

      var header = window.document.querySelector("#quarto-embed-header");
      const titleBannerEl = window.document.querySelector("body > #title-block-header");
      if (titleBannerEl) {
        titleBannerEl.style.paddingTop = header.clientHeight + "px";
      }
      const contentEl = window.document.getElementById('quarto-content');
      for (const child of contentEl.children) {
        child.style.paddingTop = header.clientHeight + "px";
        child.style.marginTop = "1em";
      }

      // Use the article root if the `back` call doesn't work. This isn't perfect
      // but should typically work
      window.quartoBackToArticle = () => {
        var currentUrl = window.location.href;
        window.history.back();
        setTimeout(() => {
            // if location was not changed in 100 ms, then there is no history back
            if(currentUrl === window.location.href){              
                // redirect to site root
                window.location.href = "index.html";
            }
        }, 100);
      }

      const headroom = new window.Headroom(header, {
        tolerance: 5,
        onPin: function () {
        },
        onUnpin: function () {
        },
      });
      headroom.init();
    });
    </script>

    
<script src="site_libs/manuscript-notebook/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
     <script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>   <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script> 
      <meta name="citation_title" content="Harnessing generative AI to annotate the severity of all phenotypic abnormalities within the Human Phenotype Ontology">
<meta name="citation_keywords" content="Ontology,Human Phenotype Ontology,Large Language Model,GPT-4,Artificial intelligence,Generative AI,Rare disease,Automation,Medicine">
<meta name="citation_author" content="Kitty B Murphy">
<meta name="citation_author" content="Brian M Schilder">
<meta name="citation_author" content="Nathan G Skene">
<meta name="citation_publication_date" content="2024-05-22">
<meta name="citation_cover_date" content="2024-05-22">
<meta name="citation_year" content="2024">
<meta name="citation_online_date" content="2024-05-22">
<meta name="citation_language" content="en">
<meta name="citation_journal_title" content="medRxiv">
<meta name="citation_reference" content="citation_title=The human phenotype ontology in 2024: Phenotypes around the world;,citation_abstract=The Human Phenotype Ontology (HPO) is a widely used resource that comprehensively organizes and defines the phenotypic features of human disease, enabling computational inference and supporting genomic and phenotypic analyses through semantic similarity and machine learning algorithms. The HPO has widespread applications in clinical diagnostics and translational research, including genomic diagnostics, gene-disease discovery, and cohort analytics. In recent years, groups around the world have developed translations of the HPO from English to other languages, and the HPO browser has been internationalized, allowing users to view HPO term labels and in many cases synonyms and definitions in ten languages in addition to English. Since our last report, a total of 2239 new HPO terms and 49235 new HPO annotations were developed, many in collaboration with external groups in the fields of psychiatry, arthrogryposis, immunology and cardiology. The Medical Action Ontology (MAxO) is a new effort to model treatments and other measures taken for clinical management. Finally, the HPO consortium is contributing to efforts to integrate the HPO and the GA4GH Phenopacket Schema into electronic health records (EHRs) with the goal of more standardized and computable integration of rare disease data in EHRs.;,citation_author=Michael A Gargano;,citation_author=Nicolas Matentzoglu;,citation_author=Ben Coleman;,citation_author=Eunice B Addo-Lartey;,citation_author=Anna V Anagnostopoulos;,citation_author=Joel Anderton;,citation_author=Paul Avillach;,citation_author=Anita M Bagley;,citation_author=Eduard Bakštein;,citation_author=James P Balhoff;,citation_author=Gareth Baynam;,citation_author=Susan M Bello;,citation_author=Michael Berk;,citation_author=Holli Bertram;,citation_author=Somer Bishop;,citation_author=Hannah Blau;,citation_author=David F Bodenstein;,citation_author=Pablo Botas;,citation_author=Kaan Boztug;,citation_author=Jolana Čady;,citation_author=Tiffany J Callahan;,citation_author=Rhiannon Cameron;,citation_author=Seth J Carbon;,citation_author=Francisco Castellanos;,citation_author=J Harry Caufield;,citation_author=Lauren E Chan;,citation_author=Christopher G Chute;,citation_author=Jaime Cruz-Rojo;,citation_author=Noémi Dahan-Oliel;,citation_author=Jon R Davids;,citation_author=Maud Dieuleveult;,citation_author=Vinicius Souza;,citation_author=Bert B A Vries;,citation_author=Esther Vries;,citation_author=J Raymond DePaulo;,citation_author=Beata Derfalvi;,citation_author=Ferdinand Dhombres;,citation_author=Claudia Diaz-Byrd;,citation_author=Alexander J M Dingemans;,citation_author=Bruno Donadille;,citation_author=Michael Duyzend;,citation_author=Reem Elfeky;,citation_author=Shahim Essaid;,citation_author=Carolina Fabrizzi;,citation_author=Giovanna Fico;,citation_author=Helen V Firth;,citation_author=Yun Freudenberg-Hua;,citation_author=Janice M Fullerton;,citation_author=Davera L Gabriel;,citation_author=Kimberly Gilmour;,citation_author=Jessica Giordano;,citation_author=Fernando S Goes;,citation_author=Rachel Gore Moses;,citation_author=Ian Green;,citation_author=Matthias Griese;,citation_author=Tudor Groza;,citation_author=Weihong Gu;,citation_author=Julia Guthrie;,citation_author=Benjamin Gyori;,citation_author=Ada Hamosh;,citation_author=Marc Hanauer;,citation_author=Kateřina Hanušová;,citation_author=Yongqun Oliver He;,citation_author=Harshad Hegde;,citation_author=Ingo Helbig;,citation_author=Kateřina Holasová;,citation_author=Charles Tapley Hoyt;,citation_author=Shangzhi Huang;,citation_author=Eric Hurwitz;,citation_author=Julius O B Jacobsen;,citation_author=Xiaofeng Jiang;,citation_author=Lisa Joseph;,citation_author=Kamyar Keramatian;,citation_author=Bryan King;,citation_author=Katrin Knoflach;,citation_author=David A Koolen;,citation_author=Megan L Kraus;,citation_author=Carlo Kroll;,citation_author=Maaike Kusters;,citation_author=Markus S Ladewig;,citation_author=David Lagorce;,citation_author=Meng-Chuan Lai;,citation_author=Pablo Lapunzina;,citation_author=Bryan Laraway;,citation_author=David Lewis-Smith;,citation_author=Xiarong Li;,citation_author=Caterina Lucano;,citation_author=Marzieh Majd;,citation_author=Mary L Marazita;,citation_author=Victor Martinez-Glez;,citation_author=Toby H McHenry;,citation_author=Melvin G McInnis;,citation_author=Julie A McMurry;,citation_author=Michaela Mihulová;,citation_author=Caitlin E Millett;,citation_author=Philip B Mitchell;,citation_author=Veronika Moslerová;,citation_author=Kenji Narutomi;,citation_author=Shahrzad Nematollahi;,citation_author=Julian Nevado;,citation_author=Andrew A Nierenberg;,citation_author=Nikola Novák Čajbiková;,citation_author=John I Nurnberger;,citation_author=Soichi Ogishima;,citation_author=Daniel Olson;,citation_author=Abigail Ortiz;,citation_author=Harry Pachajoa;,citation_author=Guiomar Perez de Nanclares;,citation_author=Amy Peters;,citation_author=Tim Putman;,citation_author=Christina K Rapp;,citation_author=Ana Rath;,citation_author=Justin Reese;,citation_author=Lauren Rekerle;,citation_author=Angharad M Roberts;,citation_author=Suzy Roy;,citation_author=Stephan J Sanders;,citation_author=Catharina Schuetz;,citation_author=Eva C Schulte;,citation_author=Thomas G Schulze;,citation_author=Martin Schwarz;,citation_author=Katie Scott;,citation_author=Dominik Seelow;,citation_author=Berthold Seitz;,citation_author=Yiping Shen;,citation_author=Morgan N Similuk;,citation_author=Eric S Simon;,citation_author=Balwinder Singh;,citation_author=Damian Smedley;,citation_author=Cynthia L Smith;,citation_author=Jake T Smolinsky;,citation_author=Sarah Sperry;,citation_author=Elizabeth Stafford;,citation_author=Ray Stefancsik;,citation_author=Robin Steinhaus;,citation_author=Rebecca Strawbridge;,citation_author=Jagadish Chandrabose Sundaramurthi;,citation_author=Polina Talapova;,citation_author=Jair A Tenorio Castano;,citation_author=Pavel Tesner;,citation_author=Rhys H Thomas;,citation_author=Audrey Thurm;,citation_author=Marek Turnovec;,citation_author=Marielle E Gijn;,citation_author=Nicole A Vasilevsky;,citation_author=Markéta Vlčková;,citation_author=Anita Walden;,citation_author=Kai Wang;,citation_author=Ron Wapner;,citation_author=James S Ware;,citation_author=Addo A Wiafe;,citation_author=Samuel A Wiafe;,citation_author=Lisa D Wiggins;,citation_author=Andrew E Williams;,citation_author=Chen Wu;,citation_author=Margot J Wyrwoll;,citation_author=Hui Xiong;,citation_author=Nefize Yalin;,citation_author=Yasunori Yamamoto;,citation_author=Lakshmi N Yatham;,citation_author=Anastasia K Yocum;,citation_author=Allan H Young;,citation_author=Zafer Yüksel;,citation_author=Peter P Zandi;,citation_author=Andreas Zankl;,citation_author=Ignacio Zarante;,citation_author=Miroslav Zvolský;,citation_author=Sabrina Toro;,citation_author=Leigh C Carmody;,citation_author=Nomi L Harris;,citation_author=Monica C Munoz-Torres;,citation_author=Daniel Danis;,citation_author=Christopher J Mungall;,citation_author=Sebastian Köhler;,citation_author=Melissa A Haendel;,citation_author=Peter N Robinson;,citation_publication_date=2024-01;,citation_cover_date=2024-01;,citation_year=2024;,citation_issue=D1;,citation_volume=52;,citation_journal_title=Nucleic Acids Res.;">
<meta name="citation_reference" content="citation_title=Systematic classification of disease severity for evaluation of expanded carrier screening panels;,citation_abstract=Professional guidelines dictate that disease severity is a key criterion for carrier screening. Expanded carrier screening, which tests for hundreds to thousands of mutations simultaneously, requires an objective, systematic means of describing a given disease’s severity to build screening panels. We hypothesized that diseases with characteristics deemed to be of highest impact would likewise be rated as most severe, and diseases with characteristics of lower impact would be rated as less severe. We describe a pilot test of this hypothesis in which we surveyed 192 health care professionals to determine the impact of specific disease phenotypic characteristics on perceived severity, and asked the same group to rate the severity of selected inherited diseases. The results support the hypothesis: we identified four “Tiers” of disease characteristics (1-4). Based on these responses, we developed an algorithm that, based on the combination of characteristics normally seen in an affected individual, classifies the disease as Profound, Severe, Moderate, or Mild. This algorithm allows simple classification of disease severity that is replicable and not labor intensive.;,citation_author=Gabriel A Lazarin;,citation_author=Felicia Hawthorne;,citation_author=Nicholas S Collins;,citation_author=Elizabeth A Platt;,citation_author=Eric A Evans;,citation_author=Imran S Haque;,citation_publication_date=2014-12;,citation_cover_date=2014-12;,citation_year=2014;,citation_issue=12;,citation_volume=9;,citation_journal_title=PLoS One;">
<meta name="citation_reference" content="citation_title=GPT-4 technical report;,citation_author=undefined OpenAI;,citation_author=Josh Achiam;,citation_author=Steven Adler;,citation_author=Sandhini Agarwal;,citation_author=Lama Ahmad;,citation_author=Ilge Akkaya;,citation_author=Florencia Leoni Aleman;,citation_author=Diogo Almeida;,citation_author=Janko Altenschmidt;,citation_author=Sam Altman;,citation_author=undefined al.;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=http://arxiv.org/abs/2303.08774;,citation_doi=10.48550/arXiv.2303.08774;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Adapted large language models can outperform medical experts in clinical text summarization;,citation_author=Dave Van Veen;,citation_author=Cara Van Uden;,citation_author=Louis Blankemeier;,citation_author=Jean-Benoit Delbrouck;,citation_author=Asad Aali;,citation_author=Christian Bluethgen;,citation_author=Anuj Pareek;,citation_author=Malgorzata Polacin;,citation_author=Eduardo Pontes Reis;,citation_author=Anna Seehofnerová;,citation_author=undefined al.;,citation_publication_date=2024-02;,citation_cover_date=2024-02;,citation_year=2024;,citation_fulltext_html_url=https://www.nature.com/articles/s41591-024-02855-5;,citation_doi=10.1038/s41591-024-02855-5;,citation_issn=1546-170X;,citation_journal_title=Nature Medicine;,citation_publisher=Nature;">
<meta name="citation_reference" content="citation_title=Large language models encode clinical knowledge;,citation_author=Karan Singhal;,citation_author=Shekoofeh Azizi;,citation_author=Tao Tu;,citation_author=S. Sara Mahdavi;,citation_author=Jason Wei;,citation_author=Hyung Won Chung;,citation_author=Nathan Scales;,citation_author=Ajay Tanwani;,citation_author=Heather Cole-Lewis;,citation_author=Stephen Pfohl;,citation_author=undefined al.;,citation_publication_date=2023-07;,citation_cover_date=2023-07;,citation_year=2023;,citation_fulltext_html_url=https://www.nature.com/articles/s41586-023-06291-2;,citation_doi=10.1038/s41586-023-06291-2;,citation_issn=1476-4687;,citation_journal_title=Nature;,citation_publisher=Nature;">
<meta name="citation_reference" content="citation_title=BioGPT: Generative pre-trained transformer for biomedical text generation and mining;,citation_author=Renqian Luo;,citation_author=Liai Sun;,citation_author=Yingce Xia;,citation_author=Tao Qin;,citation_author=Sheng Zhang;,citation_author=Hoifung Poon;,citation_author=Tie-Yan Liu;,citation_publication_date=2022-11;,citation_cover_date=2022-11;,citation_year=2022;,citation_fulltext_html_url=https://doi.org/10.1093/bib/bbac409;,citation_issue=6;,citation_doi=10.1093/bib/bbac409;,citation_issn=1477-4054;,citation_volume=23;,citation_journal_title=Briefings in Bioinformatics;">
<meta name="citation_reference" content="citation_title=Exploring the potential of GPT-4 in biomedical engineering: The dawn of a new era;,citation_author=Kunming Cheng;,citation_author=Qiang Guo;,citation_author=Yongbin He;,citation_author=Yanqiu Lu;,citation_author=Shuqin Gu;,citation_author=Haiyang Wu;,citation_publication_date=2023-08;,citation_cover_date=2023-08;,citation_year=2023;,citation_issue=8;,citation_doi=10.1007/s10439-023-03221-1;,citation_issn=1573-9686;,citation_volume=51;,citation_journal_title=Annals of Biomedical Engineering;">
<meta name="citation_reference" content="citation_title=BiomedGPT: A unified and generalist biomedical generative pre-trained transformer for vision, language, and multimodal tasks;,citation_author=Kai Zhang;,citation_author=Jun Yu;,citation_author=Zhiling Yan;,citation_author=Yixin Liu;,citation_author=Eashan Adhikarla;,citation_author=Sunyang Fu;,citation_author=Xun Chen;,citation_author=Chen Chen;,citation_author=Yuyin Zhou;,citation_author=Xiang Li;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2305.17100;,citation_doi=10.48550/arXiv.2305.17100;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=BioMedLM: A 2.7B parameter language model trained on biomedical text;,citation_author=Elliot Bolton;,citation_author=Abhinav Venigalla;,citation_author=Michihiro Yasunaga;,citation_author=David Hall;,citation_author=Betty Xiong;,citation_author=Tony Lee;,citation_author=Roxana Daneshjou;,citation_author=Jonathan Frankle;,citation_author=Percy Liang;,citation_author=Michael Carbin;,citation_author=undefined al.;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=http://arxiv.org/abs/2403.18421;,citation_doi=10.48550/arXiv.2403.18421;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Domain-specific language model pretraining for biomedical natural language processing;,citation_author=Yu Gu;,citation_author=Robert Tinn;,citation_author=Hao Cheng;,citation_author=Michael Lucas;,citation_author=Naoto Usuyama;,citation_author=Xiaodong Liu;,citation_author=Tristan Naumann;,citation_author=Jianfeng Gao;,citation_author=Hoifung Poon;,citation_publication_date=2021-10;,citation_cover_date=2021-10;,citation_year=2021;,citation_fulltext_html_url=https://dl.acm.org/doi/10.1145/3458754;,citation_issue=1;,citation_doi=10.1145/3458754;,citation_volume=3;,citation_journal_title=ACM Transactions on Computing for Healthcare;">
<meta name="citation_reference" content="citation_title=BioMegatron: Larger biomedical domain language model;,citation_author=Hoo-Chang Shin;,citation_author=Yang Zhang;,citation_author=Evelina Bakhturina;,citation_author=Raul Puri;,citation_author=Mostofa Patwary;,citation_author=Mohammad Shoeybi;,citation_author=Raghav Mani;,citation_editor=Bonnie Webber;,citation_editor=Trevor Cohn;,citation_editor=Yulan He;,citation_editor=Yang Liu;,citation_publication_date=2020-11;,citation_cover_date=2020-11;,citation_year=2020;,citation_fulltext_html_url=https://aclanthology.org/2020.emnlp-main.379;,citation_doi=10.18653/v1/2020.emnlp-main.379;,citation_conference_title=Proceedings of the 2020 conference on empirical methods in natural language processing (EMNLP);,citation_conference=Association for Computational Linguistics;">
<meta name="citation_reference" content="citation_title=Towards accurate differential diagnosis with large language models;,citation_author=Daniel McDuff;,citation_author=Mike Schaekermann;,citation_author=Tao Tu;,citation_author=Anil Palepu;,citation_author=Amy Wang;,citation_author=Jake Garrison;,citation_author=Karan Singhal;,citation_author=Yash Sharma;,citation_author=Shekoofeh Azizi;,citation_author=Kavita Kulkarni;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2312.00164;,citation_doi=10.48550/arXiv.2312.00164;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Towards expert-level medical question answering with large language models;,citation_author=Karan Singhal;,citation_author=Tao Tu;,citation_author=Juraj Gottweis;,citation_author=Rory Sayres;,citation_author=Ellery Wulczyn;,citation_author=Le Hou;,citation_author=Kevin Clark;,citation_author=Stephen Pfohl;,citation_author=Heather Cole-Lewis;,citation_author=Darlene Neal;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2305.09617;,citation_doi=10.48550/arXiv.2305.09617;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Phenomics assistant: An interface for LLM-based biomedical knowledge graph exploration;,citation_author=Shawn T. O’Neil;,citation_author=Kevin Schaper;,citation_author=Glass Elsarboukh;,citation_author=Justin T. Reese;,citation_author=Sierra A. T. Moxon;,citation_author=Nomi L. Harris;,citation_author=Monica C. Munoz-Torres;,citation_author=Peter N. Robinson;,citation_author=Melissa A. Haendel;,citation_author=Christopher J. Mungall;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=https://www.biorxiv.org/content/10.1101/2024.01.31.578275v1;,citation_doi=10.1101/2024.01.31.578275;,citation_journal_title=bioRxiv;">
<meta name="citation_reference" content="citation_title=BioMistral: A collection of open-source pretrained large language models for medical domains;,citation_author=Yanis Labrak;,citation_author=Adrien Bazoge;,citation_author=Emmanuel Morin;,citation_author=Pierre-Antoine Gourraud;,citation_author=Mickael Rouvier;,citation_author=Richard Dufour;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=http://arxiv.org/abs/2402.10373;,citation_doi=10.48550/arXiv.2402.10373;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Can generalist foundation models outcompete special-purpose tuning? Case study in medicine;,citation_author=Harsha Nori;,citation_author=Yin Tat Lee;,citation_author=Sheng Zhang;,citation_author=Dean Carignan;,citation_author=Richard Edgar;,citation_author=Nicolo Fusi;,citation_author=Nicholas King;,citation_author=Jonathan Larson;,citation_author=Yuanzhi Li;,citation_author=Weishung Liu;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2311.16452;,citation_doi=10.48550/arXiv.2311.16452;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Syndromes associated with syndactyly;,citation_author=Lorenzo Garagnani;,citation_author=Gillian D. Smith;,citation_editor=Joshua M. Abzug;,citation_editor=Scott Kozin;,citation_editor=Dan A. Zlotolow;,citation_publication_date=2013;,citation_cover_date=2013;,citation_year=2013;,citation_fulltext_html_url=https://doi.org/10.1007/978-1-4614-8758-6_14-1;,citation_doi=10.1007/978-1-4614-8758-6_14-1;,citation_isbn=978-1-4614-8758-6;,citation_inbook_title=The pediatric upper extremity;">
<meta name="citation_reference" content="citation_title=Molar tooth sign of the midbrain–hindbrain junction: Occurrence in multiple distinct syndromes;,citation_author=Joseph G. Gleeson;,citation_author=Lesley C. Keeler;,citation_author=Melissa A. Parisi;,citation_author=Sarah E. Marsh;,citation_author=Phillip F. Chance;,citation_author=Ian A. Glass;,citation_author=John M. Graham Jr;,citation_author=Bernard L. Maria;,citation_author=A. James Barkovich;,citation_author=William B. Dobyns;,citation_publication_date=2004;,citation_cover_date=2004;,citation_year=2004;,citation_fulltext_html_url=https://onlinelibrary.wiley.com/doi/abs/10.1002/ajmg.a.20437;,citation_issue=2;,citation_doi=10.1002/ajmg.a.20437;,citation_issn=1552-4833;,citation_volume=125A;,citation_journal_title=American Journal of Medical Genetics Part A;">
<meta name="citation_reference" content="citation_title=The human phenotype ontology in 2021;,citation_author=Sebastian Köhler;,citation_author=Michael Gargano;,citation_author=Nicolas Matentzoglu;,citation_author=Leigh C Carmody;,citation_author=David Lewis-Smith;,citation_author=Nicole A Vasilevsky;,citation_author=Daniel Danis;,citation_author=Ganna Balagura;,citation_author=Gareth Baynam;,citation_author=Amy M Brower;,citation_author=undefined al.;,citation_publication_date=2021-01;,citation_cover_date=2021-01;,citation_year=2021;,citation_fulltext_html_url=https://doi.org/10.1093/nar/gkaa1043;,citation_issue=D1;,citation_doi=10.1093/nar/gkaa1043;,citation_issn=0305-1048;,citation_volume=49;,citation_journal_title=Nucleic Acids Research;">
<meta name="citation_reference" content="citation_title=The human phenotype ontology in 2024: Phenotypes around the world;,citation_author=Michael A Gargano;,citation_author=Nicolas Matentzoglu;,citation_author=Ben Coleman;,citation_author=Eunice B Addo-Lartey;,citation_author=Anna V Anagnostopoulos;,citation_author=Joel Anderton;,citation_author=Paul Avillach;,citation_author=Anita M Bagley;,citation_author=Eduard Bakštein;,citation_author=James P Balhoff;,citation_author=undefined al.;,citation_publication_date=2024-01;,citation_cover_date=2024-01;,citation_year=2024;,citation_fulltext_html_url=https://doi.org/10.1093/nar/gkad1005;,citation_issue=D1;,citation_doi=10.1093/nar/gkad1005;,citation_issn=0305-1048;,citation_volume=52;,citation_journal_title=Nucleic Acids Research;">
<meta name="citation_reference" content="citation_title=Structured prompt interrogation and recursive extraction of semantics (SPIRES): A method for populating knowledge bases using zero-shot learning;,citation_author=J. Harry Caufield;,citation_author=Harshad Hegde;,citation_author=Vincent Emonet;,citation_author=Nomi L. Harris;,citation_author=Marcin P. Joachimiak;,citation_author=Nicolas Matentzoglu;,citation_author=HyeongSik Kim;,citation_author=Sierra A. T. Moxon;,citation_author=Justin T. Reese;,citation_author=Melissa A. Haendel;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2304.02711;,citation_doi=10.48550/arXiv.2304.02711;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Large language models and knowledge graphs: Opportunities and challenges;,citation_author=Jeff Z. Pan;,citation_author=Simon Razniewski;,citation_author=Jan-Christoph Kalo;,citation_author=Sneha Singhania;,citation_author=Jiaoyan Chen;,citation_author=Stefan Dietze;,citation_author=Hajira Jabeen;,citation_author=Janna Omeliyanenko;,citation_author=Wen Zhang;,citation_author=Matteo Lissandrini;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2308.06374;,citation_doi=10.48550/arXiv.2308.06374;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=The monarch initiative in 2024: An analytic platform integrating phenotypes, genes&nbsp;and diseases across species;,citation_author=Tim E Putman;,citation_author=Kevin Schaper;,citation_author=Nicolas Matentzoglu;,citation_author=Vincent P Rubinetti;,citation_author=Faisal S Alquaddoomi;,citation_author=Corey Cox;,citation_author=J Harry Caufield;,citation_author=Glass Elsarboukh;,citation_author=Sarah Gehrke;,citation_author=Harshad Hegde;,citation_author=undefined al.;,citation_publication_date=2024-01;,citation_cover_date=2024-01;,citation_year=2024;,citation_fulltext_html_url=https://europepmc.org/articles/PMC10767791;,citation_issue=D1;,citation_doi=10.1093/nar/gkad1082;,citation_issn=1362-4962;,citation_volume=52;,citation_journal_title=Nucleic acids research;">
<meta name="citation_reference" content="citation_title=The monarch initiative: An integrative data and analytic platform connecting phenotypes to genotypes across species;,citation_author=Christopher J. Mungall;,citation_author=Julie A. McMurry;,citation_author=Sebastian Köhler;,citation_author=James P. Balhoff;,citation_author=Charles Borromeo;,citation_author=Matthew Brush;,citation_author=Seth Carbon;,citation_author=Tom Conlin;,citation_author=Nathan Dunn;,citation_author=Mark Engelstad;,citation_author=undefined al.;,citation_publication_date=2017-01;,citation_cover_date=2017-01;,citation_year=2017;,citation_fulltext_html_url=https://doi.org/10.1093/nar/gkw1128;,citation_issue=D1;,citation_doi=10.1093/nar/gkw1128;,citation_issn=0305-1048;,citation_volume=45;,citation_journal_title=Nucleic Acids Research;">
<meta name="citation_reference" content="citation_title=Open targets platform: Supporting systematic drug–target identification and prioritisation;,citation_author=David Ochoa;,citation_author=Andrew Hercules;,citation_author=Miguel Carmona;,citation_author=Daniel Suveges;,citation_author=Asier Gonzalez-Uriarte;,citation_author=Cinzia Malangone;,citation_author=Alfredo Miranda;,citation_author=Luca Fumis;,citation_author=Denise Carvalho-Silva;,citation_author=Michaela Spitzer;,citation_author=undefined al.;,citation_publication_date=2021-01;,citation_cover_date=2021-01;,citation_year=2021;,citation_fulltext_html_url=https://doi.org/10.1093/nar/gkaa1027;,citation_issue=D1;,citation_doi=10.1093/nar/gkaa1027;,citation_issn=0305-1048;,citation_volume=49;,citation_journal_title=Nucleic Acids Research;">
<meta name="citation_reference" content="citation_title=Dynamic retrieval augmented generation of ontologies using artificial intelligence (DRAGON-AI);,citation_author=Sabrina Toro;,citation_author=Anna V. Anagnostopoulos;,citation_author=Sue Bello;,citation_author=Kai Blumberg;,citation_author=Rhiannon Cameron;,citation_author=Leigh Carmody;,citation_author=Alexander D. Diehl;,citation_author=Damion Dooley;,citation_author=William Duncan;,citation_author=Petra Fey;,citation_author=undefined al.;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2312.10904;,citation_doi=10.48550/arXiv.2312.10904;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=A statistical interpretation of term specificity and its application in retrieval;,citation_author=Karen Spärck Jones;,citation_publication_date=1972-01;,citation_cover_date=1972-01;,citation_year=1972;,citation_fulltext_html_url=https://doi.org/10.1108/eb026526;,citation_issue=1;,citation_doi=10.1108/eb026526;,citation_issn=0022-0418;,citation_volume=28;,citation_journal_title=Journal of Documentation;,citation_publisher=MCB;">
<meta name="citation_reference" content="citation_title=Efficient estimation of word representations in vector space;,citation_author=Tomas Mikolov;,citation_author=Kai Chen;,citation_author=Greg Corrado;,citation_author=Jeffrey Dean;,citation_publication_date=2013;,citation_cover_date=2013;,citation_year=2013;,citation_fulltext_html_url=http://arxiv.org/abs/1301.3781;,citation_doi=10.48550/arXiv.1301.3781;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Long-form factuality in large language models;,citation_author=Jerry Wei;,citation_author=Chengrun Yang;,citation_author=Xinying Song;,citation_author=Yifeng Lu;,citation_author=Nathan Hu;,citation_author=Jie Huang;,citation_author=Dustin Tran;,citation_author=Daiyi Peng;,citation_author=Ruibo Liu;,citation_author=Da Huang;,citation_author=undefined al.;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=http://arxiv.org/abs/2403.18802;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Aspects of human memory and large language models;,citation_author=Romuald A. Janik;,citation_publication_date=2024;,citation_cover_date=2024;,citation_year=2024;,citation_fulltext_html_url=http://arxiv.org/abs/2311.03839;,citation_doi=10.48550/arXiv.2311.03839;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Interpretability and explainability: A machine learning zoo mini-tour;,citation_author=Ričards Marcinkevičs;,citation_author=Julia E. Vogt;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_fulltext_html_url=http://arxiv.org/abs/2012.01805;,citation_doi=10.48550/arXiv.2012.01805;,citation_journal_title=arXiv;">
<meta name="citation_reference" content="citation_title=Identification of cell type-specific gene targets underlying thousands of rare diseases and subtraits;,citation_author=Kitty B. Murphy;,citation_author=Robert Gordon-Smith;,citation_author=Jai Chapman;,citation_author=Momoko Otani;,citation_author=Brian M. Schilder;,citation_author=Nathan G. Skene;,citation_publication_date=2023;,citation_cover_date=2023;,citation_year=2023;,citation_doi=10.1101/2023.02.13.23285820;,citation_language=en;,citation_journal_title=medRxiv;">
<meta name="citation_reference" content="citation_title=Problems and solutions in calculating quality-adjusted life years (QALYs);,citation_author=Luis Prieto;,citation_author=José A Sacristán;,citation_publication_date=2003-12;,citation_cover_date=2003-12;,citation_year=2003;,citation_fulltext_html_url=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC317370/;,citation_doi=10.1186/1477-7525-1-80;,citation_issn=1477-7525;,citation_volume=1;,citation_journal_title=Health and Quality of Life Outcomes;">
<meta name="citation_reference" content="citation_title=Prenatal diagnosis and genetic analysis of fetal akinesia deformation sequence and multiple pterygium syndrome associated with neuromuscular junction disorders: A review;,citation_author=Chih-Ping Chen;,citation_publication_date=2012-03;,citation_cover_date=2012-03;,citation_year=2012;,citation_fulltext_html_url=https://www.sciencedirect.com/science/article/pii/S1028455912000058;,citation_issue=1;,citation_doi=10.1016/j.tjog.2012.01.004;,citation_issn=1028-4559;,citation_volume=51;,citation_journal_title=Taiwanese Journal of Obstetrics and Gynecology;">
</head>

  <body class="quarto-notebook">
    <div id="quarto-embed-header" class="headroom fixed-top bg-primary">
      
      <a onclick="window.quartoBackToArticle(); return false;" class="btn btn-primary quarto-back-link" href=""><i class="bi bi-caret-left"></i> Back to Article</a>
      <h6><i class="bi bi-journal-code"></i> Article Notebook</h6>

            <a href="./manuscript.qmd" class="btn btn-primary quarto-download-embed" download="manuscript.qmd">Download Source</a>
          </div>

     <header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Harnessing generative AI to annotate the severity of all phenotypic abnormalities within the Human Phenotype Ontology</h1>
          </div>

    
    <div class="quarto-title-meta-container">
      <div class="quarto-title-meta-column-start">
            <div class="quarto-title-meta-author">
          <div class="quarto-title-meta-heading">Authors</div>
          <div class="quarto-title-meta-heading">Affiliations</div>
          
                <div class="quarto-title-meta-contents">
            <p class="author">Kitty B Murphy <a href="mailto:kitty.murphy@ukdri.ac.uk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-8669-3076" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Department of Brain Sciences, Imperial College London, UK
                      </p>
                    <p class="affiliation">
                        UK Dementia Research Institute at Imperial College London, UK
                      </p>
                  </div>
                      <div class="quarto-title-meta-contents">
            <p class="author">Brian M Schilder <a href="mailto:brian_schilder@alumni.brown.edu" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-5949-2191" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Department of Brain Sciences, Imperial College London, UK
                      </p>
                    <p class="affiliation">
                        UK Dementia Research Institute at Imperial College London, UK
                      </p>
                  </div>
                      <div class="quarto-title-meta-contents">
            <p class="author">Nathan G Skene <a href="mailto:n.skene@imperial.ac.uk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-6807-3180" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Department of Brain Sciences, Imperial College London, UK
                      </p>
                    <p class="affiliation">
                        UK Dementia Research Institute at Imperial College London, UK
                      </p>
                  </div>
                    </div>
        
        <div class="quarto-title-meta">

                      
                <div>
            <div class="quarto-title-meta-heading">Published</div>
            <div class="quarto-title-meta-contents">
              <p class="date">May 22, 2024</p>
            </div>
          </div>
          
                
              </div>
      </div>
      <div class="quarto-title-meta-column-end quarto-other-formats-target">
      </div>
    </div>


    <div>
      <div class="keywords">
        <div class="block-title">Keywords</div>
        <p>Ontology, Human Phenotype Ontology, Large Language Model, GPT-4, Artificial intelligence, Generative AI, Rare disease, Automation, Medicine</p>
      </div>
    </div>

    <div class="quarto-other-links-text-target">
    </div>  </div>
</header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract">Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results">Results</a>
  <ul class="collapse">
  <li><a href="#annotating-the-hpo-using-gpt-4" id="toc-annotating-the-hpo-using-gpt-4" class="nav-link" data-scroll-target="#annotating-the-hpo-using-gpt-4">Annotating the HPO using GPT-4</a></li>
  <li><a href="#annotation-consistency-and-recall" id="toc-annotation-consistency-and-recall" class="nav-link" data-scroll-target="#annotation-consistency-and-recall">Annotation consistency and recall</a></li>
  <li><a href="#quantifying-phenotypic-severity" id="toc-quantifying-phenotypic-severity" class="nav-link" data-scroll-target="#quantifying-phenotypic-severity">Quantifying phenotypic severity</a></li>
  <li><a href="#severity-classes" id="toc-severity-classes" class="nav-link" data-scroll-target="#severity-classes">Severity classes</a></li>
  <li><a href="#correlations-between-severity-metrics" id="toc-correlations-between-severity-metrics" class="nav-link" data-scroll-target="#correlations-between-severity-metrics">Correlations between severity metrics</a></li>
  <li><a href="#congenital-onset-by-hpo-branch" id="toc-congenital-onset-by-hpo-branch" class="nav-link" data-scroll-target="#congenital-onset-by-hpo-branch">Congenital onset by HPO branch</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion">Discussion</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a>
  <ul class="collapse">
  <li><a href="#annotating-the-hpo-using-openai-gpt-4" id="toc-annotating-the-hpo-using-openai-gpt-4" class="nav-link" data-scroll-target="#annotating-the-hpo-using-openai-gpt-4">Annotating the HPO using OpenAI GPT-4</a></li>
  <li><a href="#calculating-the-true-positive-rate" id="toc-calculating-the-true-positive-rate" class="nav-link" data-scroll-target="#calculating-the-true-positive-rate">Calculating the true positive rate</a></li>
  <li><a href="#quantifying-phenotypic-severity-1" id="toc-quantifying-phenotypic-severity-1" class="nav-link" data-scroll-target="#quantifying-phenotypic-severity-1">Quantifying phenotypic severity</a></li>
  <li><a href="#severity-classes-1" id="toc-severity-classes-1" class="nav-link" data-scroll-target="#severity-classes-1">Severity classes</a></li>
  <li><a href="#correlations-between-severity-metrics-1" id="toc-correlations-between-severity-metrics-1" class="nav-link" data-scroll-target="#correlations-between-severity-metrics-1">Correlations between severity metrics</a></li>
  </ul></li>
  <li><a href="#data-and-code-availability-statement" id="toc-data-and-code-availability-statement" class="nav-link" data-scroll-target="#data-and-code-availability-statement">Data and code availability statement</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements">Acknowledgements</a>
  <ul class="collapse">
  <li><a href="#funding" id="toc-funding" class="nav-link" data-scroll-target="#funding">Funding</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  <li><a href="#supplementary-materials" id="toc-supplementary-materials" class="nav-link" data-scroll-target="#supplementary-materials">Supplementary Materials</a>
  <ul class="collapse">
  <li><a href="#supplementary-figures" id="toc-supplementary-figures" class="nav-link" data-scroll-target="#supplementary-figures">Supplementary Figures</a></li>
  <li><a href="#supplementary-tables" id="toc-supplementary-tables" class="nav-link" data-scroll-target="#supplementary-tables">Supplementary Tables</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">      

       <div class="cell-container"><div class="cell-decorator"><pre>In [1]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtext)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HPOExplorer)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(KGExplorer)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'data.table'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    between, first, last</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Hmisc'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    src, summarize</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    format.pval, units</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>corrplot 0.92 loaded</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggstatsplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You can cite this package as:
     Patil, I. (2021). Visualizations with statistical details: The 'ggstatsplot' approach.
     Journal of Open Source Software, 6(61), 3167, doi:10.21105/joss.03167</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="st">"knitr.table.format"</span><span class="ot">=</span><span class="st">"pipe"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [2]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Load HPO data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>tag <span class="ot">&lt;-</span> <span class="st">"v2024-02-08"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>save_dir <span class="ot">&lt;-</span> <span class="st">"data"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>keep_descendants <span class="ot">&lt;-</span> <span class="st">"Phenotypic abnormality"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>p2g <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">load_phenotype_to_genes</span>(<span class="at">tag =</span> tag,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">save_dir=</span>save_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading cached RDS file: phenotype_to_genes.txt</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>+ Version: v2024-02-08</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>hpo <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">get_hpo</span>(<span class="at">method=</span><span class="st">"github"</span>,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">tag=</span>tag, </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">save_dir=</span>save_dir)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>hpo <span class="ot">&lt;-</span> KGExplorer<span class="sc">::</span><span class="fu">filter_ontology</span>(<span class="at">ont =</span> hpo, </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">keep_descendants =</span> keep_descendants)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Keeping descendants of 1 term(s).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>17,548 terms remain after filtering.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Read HPO GPT annotations </span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>gpt_annot <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_read</span>(<span class="at">hpo =</span> hpo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: piggyback</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading cached RDS file: phenotype_to_genes.txt</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>+ Version: v2024-04-26</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>492 phenotypes do not have matching HPO IDs.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading in GPT annotations for 16,648 phenotypes.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Compute IDs counts</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>ids_matched <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">c</span>(hpo<span class="sc">@</span>terms,<span class="fu">names</span>(hpo<span class="sc">@</span>alternative_terms)), </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                         gpt_annot<span class="sc">$</span>hpo_id)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>names_matched <span class="ot">&lt;-</span> <span class="fu">intersect</span>(hpo<span class="sc">@</span>elementMetadata<span class="sc">$</span>name,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">trimws</span>(<span class="fu">gsub</span>(<span class="st">"obsolete"</span>,<span class="st">""</span>,gpt_annot<span class="sc">$</span>hpo_name, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>n_ids_hpo <span class="ot">&lt;-</span> hpo<span class="sc">@</span>n_terms</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>n_ids_gpt <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(data.table<span class="sc">::</span><span class="fu">fcoalesce</span>(gpt_annot<span class="sc">$</span>hpo_name,gpt_annot<span class="sc">$</span>hpo_id)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [3]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>query_hits <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">search_hpo</span>(<span class="at">search_cols =</span> <span class="fu">c</span>(<span class="st">"name"</span>),</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">hpo =</span> hpo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Querying HPO for matching terms.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Number of phenotype hits per query group:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> - intellectual_disability: 19
 - impaired_mobility: 319
 - physical_malformations: 78
 - blindness: 1
 - sensory_impairments: 252
 - immunodeficiency: 29
 - cancer: 56
 - reduced_fertility: 9</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Run annotation checks</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>checks <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_check</span>(<span class="at">annot =</span> gpt_annot, </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">query_hits =</span> query_hits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `facets` argument of `facet_grid()` is deprecated as of ggplot2 2.2.0.
ℹ Please use the `rows` argument instead.
ℹ The deprecated feature was likely used in the HPOExplorer package.
  Please report the issue at
  &lt;https://github.com/neurogenomics/HPOExplorer/issues&gt;.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># checks$true_pos_rate</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [4]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>res_coded <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_codify</span>(<span class="at">annot =</span> gpt_annot)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>annot_melt <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_melt</span>(<span class="at">res_coded =</span> res_coded)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [5]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>res_class <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_class</span>(<span class="at">res_coded =</span> res_coded)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Assigning severity classes.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>fig_severity_class <span class="ot">&lt;-</span> ggstatsplot<span class="sc">::</span><span class="fu">ggbetweenstats</span>(res_class,</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">x=</span><span class="st">"severity_class"</span>, <span class="at">y=</span><span class="st">"severity_score_gpt"</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>fig_severity_class_dt <span class="ot">&lt;-</span> MSTExplorer<span class="sc">:::</span><span class="fu">get_ggstatsplot_stats</span>(fig_severity_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "atomicVector" in cache; using the first, from namespace 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'Rmpfr'</code></pre>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [6]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>gpt_annot_plot_out <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_plot</span>(</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">annot =</span> gpt_annot[hpo_name<span class="sc">!=</span>keep_descendants],</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">keep_descendants =</span> keep_descendants, </span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding level-2 ancestor to each HPO ID.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding ancestor metadata.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ancestor metadata already present. Use force_new=TRUE to overwrite.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Keeping descendants of 1 term(s).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>17,958 terms remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>166,610 associations remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Getting absolute ontology level for 18,536 IDs.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding level-2 ancestor to each HPO ID.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding ancestor metadata.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ancestor metadata already present. Use force_new=TRUE to overwrite.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Keeping descendants of 1 term(s).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>17,958 terms remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>166,610 associations remain after filtering.</code></pre>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [7]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>weights_dict <span class="ot">&lt;-</span> <span class="fu">eval</span>(<span class="fu">formals</span>(gpt_annot_codify)<span class="sc">$</span>weights_dict)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>annotation_order <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>,<span class="st">" "</span>,<span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">unlist</span>(weights_dict), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<section id="abstract" class="level2">
<h2 class="anchored" data-anchor-id="abstract">Abstract</h2>
<p>The Human Phenotype Ontology (HPO) has played a crucial role in defining, diagnosing, prognosing, and treating human diseases by providing a standardised framework for describing phenotypic abnormalities. However, documentation on the severity of each phenotype is currently very sparse, making systematic analyses and prioritisation of results challenging. With &gt;17500 phenotypic abnormalities across &gt;8600 rare diseases, manual curation of such phenotypic annotations by experts would be exceedingly labour-intensive and time-consuming. Leveraging advances in artificial intelligence, we employed the OpenAI GPT-4 large language model (LLM) to systematically annotate the severity of all phenotypic abnormalities in the HPO. First, we benchmarked the generative LLM annotations against ground-truth labels within the HPO (e.g.&nbsp;phenotypes in the ‘Cancer’ HPO branch were annotating as causing cancer by GPT-4). True positive recall rates across different annotations ranged from 89-100% (mean=96%), clearly demonstrating the ability of GPT-4 to automate the curation process with a high degree of fidelity. Using a novel approach, we developed a severity scoring system that incorporates both the nature of the phenotype outcome and the frequency of its occurrence. These severity metrics will enable efforts to systematically prioritise which human phenotypes are most detrimental to human health, and best targets for therapeutic intervention.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Comprehensive annotation of phenotypic abnormalities is invaluable for defining, diagnosing, prognosing, and treating human disease. Since 2008, the Human Phenotype Ontology (HPO) has been instrumental to this, by providing a standardised database for the description and analysis of human phenotypes <span class="citation" data-cites="kohlerHumanPhenotypeOntology2021 garganoHumanPhenotypeOntology2024">(<a href="#ref-kohlerHumanPhenotypeOntology2021" role="doc-biblioref">Köhler et al. 2021</a>; <a href="#ref-garganoHumanPhenotypeOntology2024" role="doc-biblioref">Gargano et al. 2024</a>)</span>. Through developing open community resources, the depth and breadth of the HPO has continued to expand and there are now &gt;17500 phenotypic abnormalities across &gt;8600 diseases, described. In recent years, the HPO has expanded its disease annotations so that each HPO term can have metadata including typical age of onset and frequency. In addition, there are the <em>Clinical modifier</em> and <em>Clinical course</em> subontologies, which contains terms to describe factors including severity, triggers, mortality and progression, respectively. Describing the severity-related attributes of a disease is crucial for both research and clinical care of individuals with rare diseases. This includes enhancing diagnostic capabilities, as well as prioritising and guiding gene therapy trials.</p>
<p>To date, the HPO has largely relied on manual curation by experts including clinicians, clinical geneticists and researchers. Although this approach ensures the quality and accuracy of the ontology, it is time-consuming and labour-intensive. As a result, less than 1% of terms within the HPO have metadata pertaining to their features such as time course and severity. Artificial intelligence (AI) capabilities have advanced considerably in recent years, presenting new opportunities to integrate natural language processing technologies into assisting in the curation process. Specifically, there have recently been considerable advances in large language model (LLM) and their application to biomedical problems, in some cases performing as well or better than human clinicians on standardised medical exams and patient diagnosis tasks <span class="citation" data-cites="vanveenAdaptedLargeLanguage2024 boltonBioMedLM7BParameter2024 zhangBiomedGPTUnifiedGeneralist2023 labrakBioMistralCollectionOpenSource2024 guDomainSpecificLanguageModel2021 singhalLargeLanguageModels2023 shinBioMegatronLargerBiomedical2020 chengExploringPotentialGPT42023 singhalLargeLanguageModels2023 oneilPhenomicsAssistantInterface2024 luoBioGPTGenerativePretrained2022 mcduffAccurateDifferentialDiagnosis2023 singhalExpertLevelMedicalQuestion2023">(<a href="#ref-vanveenAdaptedLargeLanguage2024" role="doc-biblioref">Van Veen et al. 2024</a>; <a href="#ref-boltonBioMedLM7BParameter2024" role="doc-biblioref">Bolton et al. 2024</a>; <a href="#ref-zhangBiomedGPTUnifiedGeneralist2023" role="doc-biblioref">Zhang et al. 2023</a>; <a href="#ref-labrakBioMistralCollectionOpenSource2024" role="doc-biblioref">Labrak et al. 2024</a>; <a href="#ref-guDomainSpecificLanguageModel2021" role="doc-biblioref">Gu et al. 2021</a>; <a href="#ref-singhalLargeLanguageModels2023" role="doc-biblioref">Singhal, Azizi, et al. 2023</a>, <a href="#ref-singhalLargeLanguageModels2023" role="doc-biblioref">2023</a>; <a href="#ref-shinBioMegatronLargerBiomedical2020" role="doc-biblioref">Shin et al. 2020</a>; <a href="#ref-chengExploringPotentialGPT42023" role="doc-biblioref">Cheng et al. 2023</a>; <a href="#ref-oneilPhenomicsAssistantInterface2024" role="doc-biblioref">O’Neil et al. 2024</a>; <a href="#ref-luoBioGPTGenerativePretrained2022" role="doc-biblioref">Luo et al. 2022</a>; <a href="#ref-mcduffAccurateDifferentialDiagnosis2023" role="doc-biblioref">McDuff et al. 2023</a>; <a href="#ref-singhalExpertLevelMedicalQuestion2023" role="doc-biblioref">Singhal, Tu, et al. 2023</a>)</span>. Recent work has demonstrated that the Generative Pre-trained Transformer 4 (GPT-4) foundation model <span class="citation" data-cites="openaiGPT4TechnicalReport2024">(<a href="#ref-openaiGPT4TechnicalReport2024" role="doc-biblioref">OpenAI et al. 2024</a>)</span>, when combined with strategic prompt engineering, can outperform even specialist LLMs that are explicitly fine-tuned for biomedical tasks <span class="citation" data-cites="noriCanGeneralistFoundation2023">(<a href="#ref-noriCanGeneralistFoundation2023" role="doc-biblioref">Nori et al. 2023</a>)</span>. In a landmark achievement, GPT-4 was the first LLM to surpass a score of 90% in the United States Medical Licensing Examination (USML) <span class="citation" data-cites="noriCanGeneralistFoundation2023">(<a href="#ref-noriCanGeneralistFoundation2023" role="doc-biblioref">Nori et al. 2023</a>)</span>.</p>
<p>Here, we have used GPT-4 to systematically annotate the severity of 17502 / 17548 (99.7%) phenotypic abnormalities within the HPO. Our severity annotation framework was adapted from previously defined criteria developed through consultation with clinicians <span class="citation" data-cites="Lazarin2014-rz">(<a href="#ref-Lazarin2014-rz" role="doc-biblioref">Lazarin et al. 2014</a>)</span>. The authors consulted 192 healthcare professionals for their opinions on the relative severity of various clinical characteristics: they used this to create a system for categorising the severity of diseases. Briefly, each healthcare professional was sent a survey asking them to first rate how important a disease characteristic was for determining disease severity, and then to rate the severity of a set of given disease. Using the responses, the authors were able to categorise clinical characteristics into 4 ‘severity tiers’. While characteristics such as shortened lifespan in infancy and intellectual disability were identified as highly severe and placed into tier 1, sensory impairment and reduced lifespan were categorised as less severe and placed into tier 4. Standardised metrics of severity allow clinicians to quickly assess the urgency of treating a given phenotype, as well as prognosing what outcomes might be expected.</p>
<p>793 phenotypes were annotated multiple times to evaluate the consistency of responses generated by GPT-4. For a subset of phenotypes with known expected annotations, true positive rates were calculated to assess recall. Additionally, based on the clinical characteristics and their occurrence, we have quantified the severity of each phenotype, providing an example of how these clinical characteristic annotations can be used to guide prioritisation of gene therapy trials. Ultimately, we hope that our resource will be of utility to those working in rare diseases, as well as the wider healthcare community.</p>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<section id="annotating-the-hpo-using-gpt-4" class="level3">
<h3 class="anchored" data-anchor-id="annotating-the-hpo-using-gpt-4">Annotating the HPO using GPT-4</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [8]:</pre></div><div id="cell-fig-occurrence" class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset to clinical characteristics   </span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>occurr_df <span class="ot">&lt;-</span> gpt_annot[,<span class="fu">c</span>(<span class="st">"hpo_name"</span>,<span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, annotation_order)),with<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>cols_of_interest <span class="ot">&lt;-</span> <span class="fu">names</span>(occurr_df)[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an empty dataframe to store the counts</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>count_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="fu">length</span>(cols_of_interest), <span class="at">ncol =</span> <span class="dv">4</span>))</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"always"</span>, <span class="st">"often"</span>, <span class="st">"rarely"</span>, <span class="st">"never"</span>)</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(count_df) <span class="ot">&lt;-</span> cols_of_interest</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Loop through each column of interest and count the occurrences of:</span></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 'always', 'often', 'rarely', 'never'</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (col <span class="cf">in</span> cols_of_interest) {</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">table</span>(occurr_df[[col]])</span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>  count_df[col, ] <span class="ot">&lt;-</span> counts[<span class="fu">match</span>(<span class="fu">colnames</span>(count_df), <span class="fu">names</span>(counts))]</span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>count_df_long <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">gather</span>(count_df,</span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a>                               <span class="at">key =</span> <span class="st">"condition"</span>,</span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a>                               <span class="at">value =</span> <span class="st">"count"</span>, </span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a>                               always<span class="sc">:</span>never)</span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a>count_df_long<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">rownames</span>(count_df)</span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a>count_df_long<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, count_df_long<span class="sc">$</span>category)</span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>count_df_long<span class="sc">$</span>condition <span class="ot">&lt;-</span> <span class="fu">factor</span>(count_df_long<span class="sc">$</span>condition, </span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"always"</span>, <span class="st">"often"</span>, <span class="st">"rarely"</span>, <span class="st">"never"</span>))</span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a>count_df_long <span class="ot">&lt;-</span> count_df_long <span class="sc">|&gt;</span></span>
<span id="cb76-27"><a href="#cb76-27" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(category) <span class="sc">|&gt;</span></span>
<span id="cb76-28"><a href="#cb76-28" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">percentage =</span> count <span class="sc">/</span> <span class="fu">sum</span>(count) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb76-29"><a href="#cb76-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-30"><a href="#cb76-30" aria-hidden="true" tabindex="-1"></a><span class="co"># set annotation order for plots </span></span>
<span id="cb76-31"><a href="#cb76-31" aria-hidden="true" tabindex="-1"></a>count_df_long<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">factor</span>(count_df_long<span class="sc">$</span>category, </span>
<span id="cb76-32"><a href="#cb76-32" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">levels =</span> annotation_order)</span>
<span id="cb76-33"><a href="#cb76-33" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb76-34"><a href="#cb76-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-35"><a href="#cb76-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the prompt generation function </span></span>
<span id="cb76-36"><a href="#cb76-36" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"code/prompt_gen.R"</span>)</span>
<span id="cb76-37"><a href="#cb76-37" aria-hidden="true" tabindex="-1"></a>prompts <span class="ot">&lt;-</span> <span class="fu">prompt_gen_table</span>(<span class="at">hpo =</span> hpo)</span>
<span id="cb76-38"><a href="#cb76-38" aria-hidden="true" tabindex="-1"></a>txt_prompt <span class="ot">&lt;-</span> prompts<span class="sc">$</span>prompt[<span class="dv">1688</span>]</span>
<span id="cb76-39"><a href="#cb76-39" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> prompts<span class="sc">$</span>terms[[<span class="dv">1688</span>]]){</span>
<span id="cb76-40"><a href="#cb76-40" aria-hidden="true" tabindex="-1"></a>  txt_prompt <span class="ot">&lt;-</span> <span class="fu">sub</span>(x,<span class="fu">paste0</span>(<span class="st">"*"</span>,x,<span class="st">"*"</span>),txt_prompt)</span>
<span id="cb76-41"><a href="#cb76-41" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb76-42"><a href="#cb76-42" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="fu">c</span>(</span>
<span id="cb76-43"><a href="#cb76-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">eval</span>(<span class="fu">formals</span>(prompt_gen_table)<span class="sc">$</span>effects),</span>
<span id="cb76-44"><a href="#cb76-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">eval</span>(<span class="fu">formals</span>(prompt_gen_table)<span class="sc">$</span>responses),</span>
<span id="cb76-45"><a href="#cb76-45" aria-hidden="true" tabindex="-1"></a>  <span class="st">"justification"</span>)){</span>
<span id="cb76-46"><a href="#cb76-46" aria-hidden="true" tabindex="-1"></a>  txt_prompt <span class="ot">&lt;-</span> <span class="fu">sub</span>(x,<span class="fu">paste0</span>(<span class="st">"**"</span>,x,<span class="st">"**"</span>),txt_prompt)</span>
<span id="cb76-47"><a href="#cb76-47" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb76-48"><a href="#cb76-48" aria-hidden="true" tabindex="-1"></a><span class="do">## Subplot a: Example prompt</span></span>
<span id="cb76-49"><a href="#cb76-49" aria-hidden="true" tabindex="-1"></a>f1a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb76-50"><a href="#cb76-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Example GPT-4 prompt"</span>) <span class="sc">+</span></span>
<span id="cb76-51"><a href="#cb76-51" aria-hidden="true" tabindex="-1"></a>  ggtext<span class="sc">::</span><span class="fu">geom_textbox</span>(<span class="fu">aes</span>(<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">7</span>),</span>
<span id="cb76-52"><a href="#cb76-52" aria-hidden="true" tabindex="-1"></a>                       <span class="at">box.padding =</span> <span class="fu">margin</span>(<span class="dv">20</span>, <span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">10</span>),</span>
<span id="cb76-53"><a href="#cb76-53" aria-hidden="true" tabindex="-1"></a>                       <span class="at">box.margin =</span> <span class="fu">margin</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>),</span>
<span id="cb76-54"><a href="#cb76-54" aria-hidden="true" tabindex="-1"></a>                       <span class="at">box.r =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">30</span>, <span class="st">"pt"</span>),</span>
<span id="cb76-55"><a href="#cb76-55" aria-hidden="true" tabindex="-1"></a>                       <span class="at">width =</span> <span class="fu">unit</span>(.<span class="dv">8</span>, <span class="st">"npc"</span>),</span>
<span id="cb76-56"><a href="#cb76-56" aria-hidden="true" tabindex="-1"></a>                       <span class="at">size=</span><span class="dv">3</span>,</span>
<span id="cb76-57"><a href="#cb76-57" aria-hidden="true" tabindex="-1"></a>                       <span class="at">halign=</span>.<span class="dv">5</span>,</span>
<span id="cb76-58"><a href="#cb76-58" aria-hidden="true" tabindex="-1"></a>                       <span class="at">fill=</span><span class="st">"grey90"</span>,</span>
<span id="cb76-59"><a href="#cb76-59" aria-hidden="true" tabindex="-1"></a>                       <span class="at">label=</span>txt_prompt) <span class="sc">+</span></span>
<span id="cb76-60"><a href="#cb76-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb76-61"><a href="#cb76-61" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb76-62"><a href="#cb76-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>))</span>
<span id="cb76-63"><a href="#cb76-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-64"><a href="#cb76-64" aria-hidden="true" tabindex="-1"></a><span class="do">## Subplot b: Stacked bar plot of clinical characteristic occurrence.</span></span>
<span id="cb76-65"><a href="#cb76-65" aria-hidden="true" tabindex="-1"></a>f1b <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(count_df_long, <span class="fu">aes</span>(<span class="at">x =</span> category, <span class="at">y =</span> count, <span class="at">fill =</span> condition)) <span class="sc">+</span></span>
<span id="cb76-66"><a href="#cb76-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb76-67"><a href="#cb76-67" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Clinical characteristic occurrence"</span>,</span>
<span id="cb76-68"><a href="#cb76-68" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span> <span class="st">"HPO phenotypes (n)"</span>, <span class="at">fill =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb76-69"><a href="#cb76-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb76-70"><a href="#cb76-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, </span>
<span id="cb76-71"><a href="#cb76-71" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">vjust=</span><span class="sc">-</span><span class="dv">1</span>),  </span>
<span id="cb76-72"><a href="#cb76-72" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>),  </span>
<span id="cb76-73"><a href="#cb76-73" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">+</span></span>
<span id="cb76-74"><a href="#cb76-74" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"GnBu"</span>, <span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb76-75"><a href="#cb76-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-76"><a href="#cb76-76" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb76-77"><a href="#cb76-77" aria-hidden="true" tabindex="-1"></a>  f1a<span class="sc">|</span> </span>
<span id="cb76-78"><a href="#cb76-78" aria-hidden="true" tabindex="-1"></a>    (f1b<span class="sc">/</span><span class="fu">plot_spacer</span>() ) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">1</span>, .<span class="dv">1</span>))</span>
<span id="cb76-79"><a href="#cb76-79" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb76-80"><a href="#cb76-80" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_layout</span>(<span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb76-81"><a href="#cb76-81" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">tag_levels =</span> letters) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-occurrence" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-occurrence-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-occurrence-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" data-glightbox="description: .lightbox-desc-1"><img src="manuscript_files/figure-html/fig-occurrence-1.png" class="img-fluid figure-img" width="960"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-occurrence-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: GPT-4 was able to annotate all human phenotypes based on whether they are always/often/rarely/never associated with different clinical characteristics. <strong>a</strong>, An example of the prompt input given to to GPT-4. The phenotypes listed in the second to last sentence (<em>italicised</em>) were changed to allow all HPO phenotypes to be annotated. <strong>b</strong>, Stacked bar plot showing the proportion of the occurrence of each clinical characteristic across all annotated HPO phenotypes. The terms shown on the x-axis are the clinical characteristics for which GPT-4 was asked to determine whether each phenotype caused them.
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>We employed the OpenAI GPT-4 model with Python to annotate 17502 terms within the HPO (v2024-02-08) <span class="citation" data-cites="kohlerHumanPhenotypeOntology2021 garganoHumanPhenotypeOntology2024">(<a href="#ref-kohlerHumanPhenotypeOntology2021" role="doc-biblioref">Köhler et al. 2021</a>; <a href="#ref-garganoHumanPhenotypeOntology2024" role="doc-biblioref">Gargano et al. 2024</a>)</span>. Our annotation framework was developed based on previously defined criteria for classifying disease severity <span class="citation" data-cites="Lazarin2014-rz">(<a href="#ref-Lazarin2014-rz" role="doc-biblioref">Lazarin et al. 2014</a>)</span>. We sought to evaluate the impact of phenotypes on factors including intellectual disability, death, impaired mobility, physical malformations, blindness, sensory impairments, immunodeficiency, cancer, reduced fertility, and congenital onset. Through prompt design we found that the performance of GPT-4 improved when we incorporated a scale associated with each effect and required a justification for each response. For each effect, we asked about the likelihood of its occurrence - whether it never, rarely, often, or always occurred.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [9]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>never_thresh <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>never_dt <span class="ot">&lt;-</span> <span class="fu">subset</span>(count_df_long, condition<span class="sc">==</span><span class="st">"never"</span> <span class="sc">&amp;</span> percentage<span class="sc">&gt;=</span>never_thresh)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>always_death <span class="ot">&lt;-</span> gpt_annot[,<span class="fu">list</span>(<span class="at">always_death=</span><span class="fu">all</span>(death<span class="sc">==</span><span class="st">"always"</span>),</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">hpo_name=</span>hpo_name[<span class="dv">1</span>]),</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>                          by<span class="ot">=</span><span class="st">"hpo_id"</span>][always_death<span class="sc">==</span><span class="cn">TRUE</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>Phenotype outcome occurrence varied across annotation categories. &gt;50% of phenotypes never caused blindness, sensory impairments, immunodeficiency, cancer, reduced fertility or intellectual disability. Only a minority of phenotypes (21.7%) never had a congenital onset, which is expected as rare disorders tend to be early onset genetic conditions <a href="#fig-occurrence" class="quarto-xref">Fig.&nbsp;1</a>.</p>
<p>Less than 1% of phenotype always directly resulted in death, such as ‘Stillbirth’, ‘Anencephaly’ and ‘Bilateral lung agenesis’. Only 71 phenotypes were annotated as always causing death, while 9707 phenotypes were annotated as often or rarely causing death. 7880 phenotypes were annotated as never causing death. This included 34 unique forms of syndactyly, a non-lethal condition that causes fused or webbed fingers (occurring 1 in 1,200–15,000 live births). While not life-threatening itself, syndactyly is a symptom of genetic disorders that can cause life-threatening cardiovascular and neurodevelopmental defects, such as Apert Syndrome <span class="citation" data-cites="garagnaniSyndromesAssociatedSyndactyly2013">(<a href="#ref-garagnaniSyndromesAssociatedSyndactyly2013" role="doc-biblioref">Garagnani and Smith 2013</a>)</span>. This example highlights the ability of GPT-4 to successfully distinguish between phenotypes that directly cause lethality, and those that are often associated with diseases that cause lethality.</p>
</section>
<section id="annotation-consistency-and-recall" class="level3">
<h3 class="anchored" data-anchor-id="annotation-consistency-and-recall">Annotation consistency and recall</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [10]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>consist <span class="ot">&lt;-</span> checks<span class="sc">$</span>annot_stringent_mean</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>consist <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">add_hpo_id</span>(consist, <span class="at">hpo =</span> hpo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>consist <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">add_ont_lvl</span>(consist, <span class="at">hpo =</span> hpo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Getting absolute ontology level for 17,548 IDs.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># consist_vs_lvl &lt;- Hmisc::rcorr(as.matrix(consist[!is.na(ontLvl)][,-c("hpo_name","hpo_id")]))</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="co"># corrplot::corrplot(consist_vs_lvl$r)</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>consist_vs_ontLvl <span class="ot">&lt;-</span> consist<span class="sc">|&gt;</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">melt.data.table</span>(<span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">"hpo_name"</span>,<span class="st">"hpo_id"</span>,<span class="st">"ontLvl"</span>), </span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">variable.name =</span> <span class="st">"metric"</span>, </span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">value.name =</span> <span class="st">"Consistent</span><span class="sc">\n</span><span class="st">(stringent)"</span>)<span class="sc">|&gt;</span></span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ggstatsplot::ggscatterstats(x = "consistency",</span></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#                             y = "ontLvl")</span></span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ggstatsplot::ggbetweenstats(x = "consistency", </span></span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>  <span class="co">#                             y = "ontLvl", type = "parametric")</span></span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>  ggstatsplot<span class="sc">::</span><span class="fu">ggbarstats</span>(<span class="at">x =</span> <span class="st">"Consistent</span><span class="sc">\n</span><span class="st">(stringent)"</span>, </span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>                          <span class="at">y =</span> <span class="st">"ontLvl"</span>) <span class="sc">+</span></span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Consistency of GPT-4 annotations by HPO level"</span>)</span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>consist_vs_ontLvl_dt <span class="ot">&lt;-</span> MSTExplorer<span class="sc">:::</span><span class="fu">get_ggstatsplot_stats</span>(consist_vs_ontLvl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>To assess annotation consistency, 793 we queried GPT-4 with a subset of the HPO phenotypes multiple times. We employed two different metrics to determine the <em>consistency rate</em>. The first, less stringent metric, defined consistency as the duplicate annotations being either ‘always’ and ‘often’, or ‘never’ and ‘rarely’. The second, more stringent metric, required exact agreement in annotation occurrences, e.g.&nbsp;‘always’ and ‘always’. For the less stringent metric, duplicated phenotypes were annotated consistently at a rate of at least 80%, and for the more stringent metric, the lowest consistency rate was 57% for congenital onset. An example of where annotations were inconsistent was for the HPO term ‘Acute leukaemia’. One time, GPT-4 annotated it as often causing impaired mobility, giving the justification that ‘weakness and fatigue from leukaemia and its treatment can impair mobility’. The other time, GPT-4 annotated it as rarely causing impaired mobility, giving the justification that ‘acute leukaemia rarely impairs mobility directly’. Despite specifying in the prompt for GPT-4 not to take into consideration indirect effects, this is an example of where it failed to do so.</p>
<p>We also reasoned that GPT-4 would be better able to give consistent answers for more specific phenotypes lower in the ontology, as they are more likely to have a single cause. We found that the stringent consistency rate did indeed significantly improve with greater HPO ontology depth (<span class="math inline">\(X_{Pearson}^2\)</span>=22.17, <span class="math inline">\(\hat{V}_{Cramer}\)</span>=0.03, p=0.05). See <a href="#fig-consist-vs-ontLvl" class="quarto-xref">Figure&nbsp;5</a> for a visual representation of this relationship.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [11]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define items to be checked</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>items <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"consistency_count"</span>, <span class="st">"consistency_rate"</span>,</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">"consistency_stringent_count"</span>, <span class="st">"consistency_stringent_rate"</span>,</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>           <span class="st">"true_pos_count"</span>, <span class="st">"true_pos_rate"</span>)</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>metric_types <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Rate"</span>, <span class="st">"Count"</span>)[<span class="dv">1</span>]</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get data frame with checks values</span></span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>check_df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(checks[items], data.table<span class="sc">::</span>as.data.table, <span class="at">keep.rownames =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">rbindlist</span>(<span class="at">idcol =</span> <span class="st">"metric"</span>) <span class="sc">|&gt;</span></span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">setnames</span>(<span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"annotation"</span>, <span class="st">"value"</span>))</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>metric <span class="ot">&lt;-</span> <span class="fu">factor</span>(check_df<span class="sc">$</span>metric,</span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> items,</span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="fu">factor</span>(check_df<span class="sc">$</span>annotation,</span>
<span id="cb83-17"><a href="#cb83-17" aria-hidden="true" tabindex="-1"></a>                              <span class="at">levels =</span> <span class="fu">unique</span>(check_df<span class="sc">$</span>annotation),</span>
<span id="cb83-18"><a href="#cb83-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb83-19"><a href="#cb83-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-20"><a href="#cb83-20" aria-hidden="true" tabindex="-1"></a>check_df[, metric_type <span class="sc">:</span><span class="er">=</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"count"</span>, metric),</span>
<span id="cb83-21"><a href="#cb83-21" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Count"</span>, <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"rate"</span>, metric), <span class="st">"Rate"</span>, <span class="cn">NA</span>))]</span>
<span id="cb83-22"><a href="#cb83-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-23"><a href="#cb83-23" aria-hidden="true" tabindex="-1"></a>check_df[, metric_category <span class="sc">:</span><span class="er">=</span> <span class="fu">gsub</span>(<span class="st">"_count|_rate"</span>, <span class="st">""</span>, metric)]</span>
<span id="cb83-24"><a href="#cb83-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-25"><a href="#cb83-25" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>metric_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(check_df<span class="sc">$</span>metric_type, <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb83-26"><a href="#cb83-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-27"><a href="#cb83-27" aria-hidden="true" tabindex="-1"></a>check_df <span class="ot">&lt;-</span> check_df[annotation <span class="sc">!=</span> <span class="st">"pheno_count"</span>]</span>
<span id="cb83-28"><a href="#cb83-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-29"><a href="#cb83-29" aria-hidden="true" tabindex="-1"></a>check_df[, n <span class="sc">:</span><span class="er">=</span> value[metric_type <span class="sc">==</span> <span class="st">"Count"</span>],</span>
<span id="cb83-30"><a href="#cb83-30" aria-hidden="true" tabindex="-1"></a>         by <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"metric_category"</span>, <span class="st">"annotation"</span>)]</span>
<span id="cb83-31"><a href="#cb83-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-32"><a href="#cb83-32" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, check_df<span class="sc">$</span>annotation)</span>
<span id="cb83-33"><a href="#cb83-33" aria-hidden="true" tabindex="-1"></a>annotation_order <span class="ot">&lt;-</span> <span class="fu">unique</span>(check_df<span class="sc">$</span>annotation)</span>
<span id="cb83-34"><a href="#cb83-34" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="fu">factor</span>(check_df<span class="sc">$</span>annotation, <span class="at">levels =</span> annotation_order)</span>
<span id="cb83-35"><a href="#cb83-35" aria-hidden="true" tabindex="-1"></a>check_df<span class="sc">$</span>metric_category <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, check_df<span class="sc">$</span>metric_category)</span>
<span id="cb83-36"><a href="#cb83-36" aria-hidden="true" tabindex="-1"></a>check_df[metric_category<span class="sc">==</span><span class="st">"consistency"</span>,metric_category<span class="sc">:</span><span class="er">=</span><span class="st">"consistency lenient"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [12]:</pre></div><div id="cell-fig-checks" class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>lab_size <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>consistency_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(check_df[metric_type <span class="sc">%in%</span> metric_types <span class="sc">&amp;</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">grepl</span>(<span class="st">"consistency"</span>,metric_category)],</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">aes</span>(<span class="at">x =</span> annotation, <span class="at">y =</span> value,</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">fill =</span> stringr<span class="sc">::</span><span class="fu">str_to_sentence</span>(</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="fu">gsub</span>(<span class="st">"consistency "</span>, <span class="st">""</span>, metric_category)</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>                                 ),</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>                               <span class="at">label =</span> <span class="fu">round</span>(value, <span class="dv">2</span>))) <span class="sc">+</span></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>,<span class="at">position =</span> <span class="st">"dodge"</span>) <span class="sc">+</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Annotation consistency"</span>,</span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span> <span class="st">"Mean consistency"</span>, </span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill=</span><span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#C3DAEAFF"</span>, <span class="st">"#ECE28BFF"</span>)) <span class="sc">+</span></span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb84-16"><a href="#cb84-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"top"</span>, </span>
<span id="cb84-17"><a href="#cb84-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">"white"</span>),</span>
<span id="cb84-18"><a href="#cb84-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb84-19"><a href="#cb84-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">1</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"n="</span>,n,<span class="st">" phenotypes"</span>)),</span>
<span id="cb84-20"><a href="#cb84-20" aria-hidden="true" tabindex="-1"></a>            <span class="at">size=</span>lab_size,</span>
<span id="cb84-21"><a href="#cb84-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">check_overlap=</span><span class="cn">TRUE</span>, <span class="at">hjust=</span><span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb84-22"><a href="#cb84-22" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Compute the empirical probability of annotations being consistent by chance </span></span>
<span id="cb84-23"><a href="#cb84-23" aria-hidden="true" tabindex="-1"></a>  <span class="do">## with 2 options and N duplicates</span></span>
<span id="cb84-24"><a href="#cb84-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">2</span><span class="sc">^</span>checks<span class="sc">$</span>annot<span class="sc">$</span>pheno_count), </span>
<span id="cb84-25"><a href="#cb84-25" aria-hidden="true" tabindex="-1"></a>             <span class="at">color=</span><span class="st">"dodgerblue"</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>)<span class="sc">+</span></span>
<span id="cb84-26"><a href="#cb84-26" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Compute the empirical probability of annotations being consistent by chance </span></span>
<span id="cb84-27"><a href="#cb84-27" aria-hidden="true" tabindex="-1"></a>  <span class="do">## with 4 options and N duplicates</span></span>
<span id="cb84-28"><a href="#cb84-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span><span class="sc">^</span>checks<span class="sc">$</span>annot<span class="sc">$</span>pheno_count), </span>
<span id="cb84-29"><a href="#cb84-29" aria-hidden="true" tabindex="-1"></a>             <span class="at">color=</span><span class="st">"gold3"</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>)</span>
<span id="cb84-30"><a href="#cb84-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-31"><a href="#cb84-31" aria-hidden="true" tabindex="-1"></a>recall_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(check_df[metric_type <span class="sc">%in%</span> metric_types <span class="sc">&amp;</span></span>
<span id="cb84-32"><a href="#cb84-32" aria-hidden="true" tabindex="-1"></a>                               metric_category <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"true pos"</span>), ],</span>
<span id="cb84-33"><a href="#cb84-33" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">aes</span>(<span class="at">x =</span> annotation, <span class="at">y =</span> value,</span>
<span id="cb84-34"><a href="#cb84-34" aria-hidden="true" tabindex="-1"></a>                            <span class="at">fill =</span> metric_category,</span>
<span id="cb84-35"><a href="#cb84-35" aria-hidden="true" tabindex="-1"></a>                            <span class="at">label =</span> n<span class="co">#paste0("n\n=\n",n)</span></span>
<span id="cb84-36"><a href="#cb84-36" aria-hidden="true" tabindex="-1"></a>                        )</span>
<span id="cb84-37"><a href="#cb84-37" aria-hidden="true" tabindex="-1"></a>                      ) <span class="sc">+</span></span>
<span id="cb84-38"><a href="#cb84-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">show.legend =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb84-39"><a href="#cb84-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb84-40"><a href="#cb84-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span> <span class="st">"Recall"</span>) <span class="sc">+</span></span>
<span id="cb84-41"><a href="#cb84-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"True positive rate"</span>) <span class="sc">+</span></span>
<span id="cb84-42"><a href="#cb84-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#B1D5BBFF"</span>)) <span class="sc">+</span></span>
<span id="cb84-43"><a href="#cb84-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb84-44"><a href="#cb84-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb84-45"><a href="#cb84-45" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb84-46"><a href="#cb84-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="at">fill =</span> <span class="fu">alpha</span>(<span class="st">"white"</span>,<span class="dv">1</span>),</span>
<span id="cb84-47"><a href="#cb84-47" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> lab_size)</span>
<span id="cb84-48"><a href="#cb84-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-49"><a href="#cb84-49" aria-hidden="true" tabindex="-1"></a>consistency_plot <span class="sc">+</span> recall_plot <span class="sc">+</span> </span>
<span id="cb84-50"><a href="#cb84-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb84-51"><a href="#cb84-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">tag_levels =</span> <span class="st">'a'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-checks" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-checks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-checks-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" data-glightbox="description: .lightbox-desc-2"><img src="manuscript_files/figure-html/fig-checks-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-checks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: GPT-4 annotations are consistent and accurate across annotations. <strong>a</strong>, Barplot showing the annotation consistency within phenotypes that were annotated more than once. In the lenient metric, annotations were collapsed into two groups (‘always’/‘often’ and ‘never’/‘rarely’). For a given clinical characteristic within a given phenotype, if an annotation was always within the same group it was considered consistent. In the stringent metric, all four annotation categories were considered to be different from one another. Thus, annotations were only defined as consistent if they were all identical. The blue dashed line indicates the probability of two annotations being consistent by chance in the lenient metric (~1/2). The gold dashed line is the probability of two annotations being consistent by chance in the stringent metric (~1/4). <strong>b</strong>, Bar plot of the true positive rate for each annotation. The labels above each bar indicate the number of phenotypes tested.
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>In order to evaluate the validity of the annotations, we calculated a true positive rate. This involved identifying specific branches within the HPO that would contain phenotypes that would reliably indicate the presence of certain conditions. For instance, the phenotypes ‘Decreased fertility in females’ and ‘Decreased fertility in males’ should often or always cause reduced fertility. We observed an encouraging true positive rate exceeding 88% across in every clinical characteristic and achieving perfect recall (100%) in 4/8 characteristics.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [13]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>lowest_recall <span class="ot">&lt;-</span> <span class="fu">names</span>(checks<span class="sc">$</span>true_pos_rate)[checks<span class="sc">$</span>true_pos_rate<span class="sc">==</span><span class="fu">min</span>(checks<span class="sc">$</span>true_pos_rate)]</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>lowest_recall_dt <span class="ot">&lt;-</span> checks<span class="sc">$</span>annot[<span class="fu">get</span>(lowest_recall)<span class="sc">==</span><span class="st">"never"</span> <span class="sc">&amp;</span> hpo_id <span class="sc">%in%</span> query_hits[[lowest_recall]]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>The lowest true positive rate was observed for physical malformations, with 88.5% recall across 87 HPO phenotypes. Some cases in which the GPT-4 annotations disagreed with the HPO ground truth included: ‘Angioma serpentinum’, ‘Nevus anemicus’, ‘Pulmonary arteriovenous fistulas’. In the case of ‘Angioma serpentinum’ it provided the justification that ‘No known association with physical malformations’. In another instance, GPT-4 noted that ‘Nevus anemicus’ is ‘Limited to hypopigmented skin patch; no other malformations.’. This indicates that while technically incorrect according to our predefined benchmarks, a case could in fact be made that mild skin conditions do not rise to the level of physical malformations.</p>
<p>This high level of recall underscores the robustness of our annotations and the reliability of the HPO framework in capturing clinically relevant phenotypic information. However, we acknowledge that the number of testable true positive phenotypes for some of these categories are low, especially ‘blindness’ for which there is only 1 phenotype in the HPO (after excluding terms pertaining to colour or night blindness). Furthermore, some of the true positive phenotypes are lexically similar to the name of the clinical characteristic itself. In these cases, annotating ‘Severe intellectual disability’ as always causing intellectual disability is a relatively trivial task. Nevertheless, even these scenarios provide a clear and interpretable benchmark for the model’s performance. In addition, were numerous phenotypes with lexically non-obvious relationships to the clinical characteristic that were annotated correctly by GPT-4. For example, ‘Molar tooth sign on MRI’ was correctly annotated as causing intellectual disability.</p>
</section>
<section id="quantifying-phenotypic-severity" class="level3">
<h3 class="anchored" data-anchor-id="quantifying-phenotypic-severity">Quantifying phenotypic severity</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [14]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>anenceph_dt <span class="ot">&lt;-</span> res_coded<span class="sc">$</span>annot_weighted[hpo_name<span class="sc">==</span><span class="st">'Anencephaly'</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>Quantifying the severity of phenotypes can have important implications for diagnosis, prognosis, and treatment, for instance, guiding prioritisation of gene therapy trials for the most severe phenotypes. First, we created a dictionary to map each phenotype outcome (e.g.&nbsp;blindness) and its response (always, often, rarely, never) to numeric values from 0-3. Then, the phenotype outcome values were multiplied by their associated response values. Importantly, the values reflected the severity of each outcome based on both the outcome itself and the frequency of the response. For instance, a phenotype always causing death would have a higher multiplied value than a phenotype rarely causing reduced fertility (see <a href="#tbl-metric-weights" class="quarto-xref">Table&nbsp;2</a>). Next, we computed an average score for each phenotype by aggregating the multiplied values across all phenotype outcomes and then calculating the mean. This was then normalised by the theoretical maximum severity score, so that all phenotypes were on a 0-100 severity scale (where 100 is the most severe phenotype possible). This average normalised score represents the overall severity of the phenotype based on the severity of its individual outcomes.</p>
<p>Based on these scores we evaluated the top 50 severe phenotypes. One of the most severe phenotype was ‘Anencephaly’ (HP:0002323) with a composite severity score of 45. Anencephaly is a birth defect where the baby is born without a portion of its brain and skull, often these babies are stillborn. In fact, many of the most severe phenotypes were related to developmental brain and neural tube defects. Comparison of the severity scores for each response, across the phenotype outcomes annotated, revealed consistent trends: as the response of the phenotype outcome increased (from never to always), the severity score also increased (<a href="#fig-severity-boxplot" class="quarto-xref">Supplementary Fig.&nbsp;7</a>). We also evaluated the severity score distribution by HPO branch and calculated the mean severity score using all phenotypes within each major HPO branch (<a href="#fig-severity-histo" class="quarto-xref">Fig.&nbsp;6</a>). The HPO branch with the greatest mean severity score was ‘Abnormal cellular phenotype’ (mean=17), followed by ‘Neoplasm’ (mean=15.2), which would include the highly ranked phenotypes seen in <a href="#fig-top-phenos" class="quarto-xref">Figure&nbsp;3</a>.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [15]:</pre></div><div id="cell-fig-top-phenos" class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>plot_top_phenos_out <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">plot_top_phenos</span>(<span class="at">res_class =</span> res_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding level-2 ancestor to each HPO ID.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding ancestor metadata.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ancestor metadata already present. Use force_new=TRUE to overwrite.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Translating ontology terms to ids.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Keeping descendants of 1 term(s).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>17,958 terms remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>166,510 associations remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Getting absolute ontology level for 18,536 IDs.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>plot_top_phenos_out<span class="sc">$</span>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-top-phenos" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-top-phenos-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-top-phenos-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" data-glightbox="description: .lightbox-desc-3"><img src="manuscript_files/figure-html/fig-top-phenos-1.png" class="img-fluid figure-img" width="1248"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-top-phenos-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Quantifying the severity of HPO phenotype annotations highlights the most impactful conditions. Heatmap of 10 represetantive phenotypes from each severity class (Profound, Severe, Moderate, Mild) stratified by whether the phenotypes are often/always congenital (<strong>a</strong>-<strong>b</strong>) or rarely/never congenital (<strong>c</strong>-<strong>d</strong>). Continuous severity scores are shown as bars (<strong>b</strong>,<strong>d</strong>) and were calculated by multiplying the numeric values assigned to each clinical characteristic according to <a href="#tbl-metric-weights" class="quarto-xref">Table&nbsp;2</a>. The average normalised score, representing overall phenotype severity on a 0-100 scale, was calculated by aggregating the multiplied values and normalising by the theoretical maximum severity score. The x-axes show each of the clinical characteristics.
</figcaption>
</figure>
</div>
</div>
</div></div>
</section>
<section id="severity-classes" class="level3">
<h3 class="anchored" data-anchor-id="severity-classes">Severity classes</h3>
<p>While the continuous severity score is a helpful metric, there may be some use cases where a categorical classification of severity is more immediately useful. In work by <span class="citation" data-cites="Lazarin2014-rz">Lazarin et al. (<a href="#ref-Lazarin2014-rz" role="doc-biblioref">2014</a>)</span>, the authors defined severity classed using a simple decision tree based on the individual severity annotations. We approximated this approach using our GPT-4 annotations. This categorical approach showed a strong degree of positive correspondence with the continuous severity score (<span class="math inline">\(\hat{\omega_{p}^2}\)</span>=0.88, p&lt;2.2e-308). In other words, severity score increased with severity class level (mild &lt; moderate &lt; severe &lt; profound) as expected. The distribution of severity classes is shown in <a href="#fig-severity-class" class="quarto-xref">Figure&nbsp;9</a>.</p>
</section>
<section id="correlations-between-severity-metrics" class="level3">
<h3 class="anchored" data-anchor-id="correlations-between-severity-metrics">Correlations between severity metrics</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [16]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>cor_metrics <span class="ot">&lt;-</span> Hmisc<span class="sc">::</span><span class="fu">rcorr</span>(<span class="fu">as.matrix</span>(res_coded<span class="sc">$</span>annot_weighted[,<span class="sc">-</span><span class="fu">c</span>(<span class="st">"hpo_id"</span>,<span class="st">"hpo_name"</span>)]))</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>cor_composite <span class="ot">&lt;-</span> <span class="fu">sort</span>(cor_metrics<span class="sc">$</span>r[<span class="st">"severity_score_gpt"</span>,], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="sc">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>We found that some severity metrics were correlated with one another, with a mean Pearson correlation of 0.2 across all individual metrics (see <a href="#fig-metric-corplot" class="quarto-xref">Figure&nbsp;8</a>). In particular, blindness and sensory impairment were highly correlated with one another (r=0.62, p=0). Some metrics drove the composite severity score more than other, which is a reflection of both our per-metric weighting scheme, response type frequencies, and the correlation structure between metrics. Overall, impaired mobility seemed to be the strongest driver of the composite severity score with a Pearson correlation of 0.6001824, followed by intellectual disability (r=0.59) and death (r=0.56).</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [17]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Correlation with QALY</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>ihme <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">"~/Downloads/IHME-GBD_2019_DATA-e678e00e-1/IHME-GBD_2019_DATA-e678e00e-1.csv"</span>)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>ihme[,nm<span class="sc">:</span><span class="er">=</span><span class="fu">tolower</span>(cause_name)]</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>res_coded<span class="sc">$</span>annot_weighted[,nm<span class="sc">:</span><span class="er">=</span><span class="fu">tolower</span>(hpo_name)]</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>ihme_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(data.table<span class="sc">::</span><span class="fu">dcast.data.table</span>(ihme, </span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">formula =</span> <span class="st">"nm~measure_name+metric_name"</span>, </span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>                             <span class="at">value.var =</span> <span class="st">"val"</span>,</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">fun.aggregate =</span> mean, <span class="at">na.rm=</span><span class="cn">TRUE</span>), </span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>                     res_coded<span class="sc">$</span>annot_weighted)</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>metric_cols <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">IHME=</span><span class="fu">grep</span>(<span class="fu">paste</span>(<span class="fu">paste0</span>(<span class="st">"_"</span>,</span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"Percent"</span>,</span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># unique(ihme$metric_name)[1],</span></span>
<span id="cb98-15"><a href="#cb98-15" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"$"</span>),<span class="at">collapse =</span> <span class="st">"|"</span>),<span class="fu">names</span>(ihme_merged), <span class="at">value =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-16"><a href="#cb98-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">GPT=</span><span class="fu">setdiff</span>(<span class="fu">names</span>(res_coded<span class="sc">$</span>annot_weighted),<span class="fu">c</span>(<span class="st">"hpo_id"</span>,<span class="st">"hpo_name"</span>,<span class="st">"nm"</span>))</span>
<span id="cb98-17"><a href="#cb98-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb98-18"><a href="#cb98-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-19"><a href="#cb98-19" aria-hidden="true" tabindex="-1"></a>Xcor <span class="ot">&lt;-</span> <span class="fu">cor</span>(ihme_merged[,<span class="fu">unlist</span>(metric_cols),<span class="at">with=</span><span class="cn">FALSE</span>], <span class="at">use=</span><span class="st">"complete.obs"</span>)</span>
<span id="cb98-20"><a href="#cb98-20" aria-hidden="true" tabindex="-1"></a>heatmaply<span class="sc">::</span><span class="fu">heatmaply</span>(Xcor[metric_cols<span class="sc">$</span>IHME,metric_cols<span class="sc">$</span>GPT])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
</section>
<section id="congenital-onset-by-hpo-branch" class="level3">
<h3 class="anchored" data-anchor-id="congenital-onset-by-hpo-branch">Congenital onset by HPO branch</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [18]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>gpt_annot_plot_branches_out <span class="ot">&lt;-</span> HPOExplorer<span class="sc">::</span><span class="fu">gpt_annot_plot_branches</span>(</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">gpt_annot =</span> gpt_annot, </span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">hpo =</span> hpo, </span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">metric =</span> <span class="st">"congenital_onset"</span>, </span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill_lab =</span> <span class="st">"Congenital</span><span class="sc">\n</span><span class="st">onset"</span>, </span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_plot =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding level-2 ancestor to each HPO ID.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Adding ancestor metadata.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ancestor metadata already present. Use force_new=TRUE to overwrite.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>17,494 associations remain after filtering.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Getting absolute ontology level for 17,548 IDs.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>top_congenital <span class="ot">&lt;-</span> <span class="fu">unique</span>(gpt_annot_plot_branches_out<span class="sc">$</span>data[<span class="sc">!</span><span class="fu">is.na</span>(congenital_onset),<span class="fu">c</span>(<span class="st">"ancestor_name"</span>,<span class="st">"proportion_always"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [19]:</pre></div><div id="cell-fig-congenital-branches" class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>gpt_annot_plot_branches_out<span class="sc">$</span>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-congenital-branches" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-congenital-branches-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-congenital-branches-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" data-glightbox="description: .lightbox-desc-4"><img src="manuscript_files/figure-html/fig-congenital-branches-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-congenital-branches-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Distribution of congenital onset across HPO branches. The y-axis shows the proportion of phenotypes that are always/often/rarely/never congenital. The x-axis shows the HPO branch, orderered from highest to lowest proportion of always congenital phenotypes.
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>Next, we assessed the distribution of congenital onset across HPO branches (<a href="#fig-congenital-branches" class="quarto-xref">Fig.&nbsp;4</a>). We found that the Abnormality of prenatal development or birth branch contained the greatest proportion of phenotypes that were always congenital (70.15%), followed by Abnormality of the musculoskeletal system (45.34%) and Growth abnormality (37.62%). This is concordant with the expectation that these phenotypes should largely be congenital. The HPO branches with the least commonly congenital phenotypes were Constitutional symptom (0%), Abnormality of the thoracic cavity (0%), and Phenotypic abnormality (0%). <a href="https://hpo.jax.org/app/browse/term/HP:0025142">‘Constitutional symptom’</a> is a fairly broad term defined as <em>‘A symptom or manifestation indicating a systemic or general effect of a disease and that may affect the general well-being or status of an individual.’</em> Examples include ‘Fatigue’ ‘Exercise intolerance’, ‘Hot flashes’ and ‘Sneeze’.</p>
</section>
</section>
<section id="discussion" class="level2">
<h2 class="anchored" data-anchor-id="discussion">Discussion</h2>
<p>The creation and annotation of biomedical knowledge has traditionally relied on manual or semi-manual curation by human experts <span class="citation" data-cites="putmanMonarchInitiative20242024 ochoaOpenTargetsPlatform2021 mungallMonarchInitiativeIntegrative2017a kohlerHumanPhenotypeOntology2021 garganoHumanPhenotypeOntology2024">(<a href="#ref-putmanMonarchInitiative20242024" role="doc-biblioref">Putman et al. 2024</a>; <a href="#ref-ochoaOpenTargetsPlatform2021" role="doc-biblioref">Ochoa et al. 2021</a>; <a href="#ref-mungallMonarchInitiativeIntegrative2017a" role="doc-biblioref">Mungall et al. 2017</a>; <a href="#ref-kohlerHumanPhenotypeOntology2021" role="doc-biblioref">Köhler et al. 2021</a>; <a href="#ref-garganoHumanPhenotypeOntology2024" role="doc-biblioref">Gargano et al. 2024</a>)</span>. Performing such manual curation and review tasks at scale is often infeasible for human biomedical experts given limited time and resources. LLMs have the capacity to effectively encode, retrieve, and synthesise vast amounts of diverse information in a highly scalable manner <span class="citation" data-cites="vanveenAdaptedLargeLanguage2024 singhalLargeLanguageModels2023 openaiGPT4TechnicalReport2024">(<a href="#ref-vanveenAdaptedLargeLanguage2024" role="doc-biblioref">Van Veen et al. 2024</a>; <a href="#ref-singhalLargeLanguageModels2023" role="doc-biblioref">Singhal, Azizi, et al. 2023</a>; <a href="#ref-openaiGPT4TechnicalReport2024" role="doc-biblioref">OpenAI et al. 2024</a>)</span>. This makes them powerful tools that can be applied in a rapidly expanding variety of scenarios, including medical practice, research and data curation <span class="citation" data-cites="singhalLargeLanguageModels2023 toroDynamicRetrievalAugmented2023 panLargeLanguageModels2023 oneilPhenomicsAssistantInterface2024 caufieldStructuredPromptInterrogation2023">(<a href="#ref-singhalLargeLanguageModels2023" role="doc-biblioref">Singhal, Azizi, et al. 2023</a>; <a href="#ref-toroDynamicRetrievalAugmented2023" role="doc-biblioref">Toro et al. 2023</a>; <a href="#ref-panLargeLanguageModels2023" role="doc-biblioref">Pan et al. 2023</a>; <a href="#ref-oneilPhenomicsAssistantInterface2024" role="doc-biblioref">O’Neil et al. 2024</a>; <a href="#ref-caufieldStructuredPromptInterrogation2023" role="doc-biblioref">Caufield et al. 2023</a>)</span>.</p>
<p>Here, we introduce a novel framework to leverage the current best-in-class LLM, GPT-4 <span class="citation" data-cites="openaiGPT4TechnicalReport2024">(<a href="#ref-openaiGPT4TechnicalReport2024" role="doc-biblioref">OpenAI et al. 2024</a>)</span>, to systematically annotate the severity of 17502 phenotypic abnormalities within the HPO. By employing advanced AI capabilities, we have demonstrated the feasibility of automating this process, significantly enhancing efficiency without substantially compromising accuracy. Our validation approach yielded a high true positive rate exceeding 88% across the phenotypes tested. Furthermore, our approach can be readily adapted and scaled to accommodate the growing volume of phenotypic data.</p>
<p>Throughout this study, we observed that GPT-4 was capable of reliably recovering deep semantic relationships from the medical domain, far beyond making superficial inferences based on lexical similarities. An excellent example of this is the phenotype ‘Molar tooth sign on MRI’ (HP:0002419; severity score=25.56), which GPT-4 annotated as causing intellectual disability. At first glance, we ourselves assumed this was a false positive as the term appeared to be related to dentition. However, upon further inspection we realised that molar tooth sign is in fact a pattern of abnormal brain morphology that happens to bear some resemblance to molar dentition when observed in radiological scans. This phenotype is a known sign of neurodevelopmental defects that can indeed cause severe intellectual disability <span class="citation" data-cites="gleesonMolarToothSign2004">(<a href="#ref-gleesonMolarToothSign2004" role="doc-biblioref">Gleeson et al. 2004</a>)</span>.</p>
<p>In addition to rapidly synthesising and summarising vast amounts of information, LLMs can also be steered to provide justifications for each particular response. This makes LLMs amenable to direct interrogation as a means of recovering explainability, especially when designed to retain information about previous requests and interactions as they use these to iteratively improve and update their predictions <span class="citation" data-cites="janikAspectsHumanMemory2024">(<a href="#ref-janikAspectsHumanMemory2024" role="doc-biblioref">Janik 2024</a>)</span>. This represents a categorical advance over traditional natural language processing models based on more shallow forms of statistical or machine learning (e.g.&nbsp;Term Frequency-Inverse Document Frequency <span class="citation" data-cites="jonesStatisticalInterpretationTerm1972">(<a href="#ref-jonesStatisticalInterpretationTerm1972" role="doc-biblioref">Jones 1972</a>)</span>, Word2vec <span class="citation" data-cites="mikolovEfficientEstimationWord2013">(<a href="#ref-mikolovEfficientEstimationWord2013" role="doc-biblioref">Mikolov et al. 2013</a>)</span>) which lack the ability to provide chains of causal reasoning to justify their predictions. This highlights the fundamental trade-off between simpler models with high explainability (the ability humans to understand the inner workings of the model) but low interpretability (the ability of humans to trace the decision process of the model, analogous to human ‘reasoning’), and deeper more complex models with low explainability but high interpretability <span class="citation" data-cites="marcinkevicsInterpretabilityExplainabilityMachine2023">(<a href="#ref-marcinkevicsInterpretabilityExplainabilityMachine2023" role="doc-biblioref">Marcinkevičs and Vogt 2023</a>)</span>.</p>
<p>A key contribution of our study is the introduction of a quantitative severity scoring system that integrates both the nature of the phenotype outcome and the frequency of its occurrence. By encoding the concept of severity in this way, we are able to prioritise phenotypes based on their impact on patients. The methodology allowed us to transition from low-throughput qualitative assessments of severity (e.g. <span class="citation" data-cites="Lazarin2014-rz">Lazarin et al. (<a href="#ref-Lazarin2014-rz" role="doc-biblioref">2014</a>)</span>) to high-throughput quantitative assessments of severity. One of the most severe phenotypes in the HPO is ‘Fetal akinesia sequence’ (FAS; HP:0001989, severity score= 43.9), and extremely rare condition that is almost always lethal. FAS is a complex, multi-system phenotype that can be caused by at least 24 different genetic disorders. Despite the complex and heterogeneous aetiology of this phenotype, GPT-4 was able to provide accurate annotations alongside explainable justifications for those annotations (see <a href="#tbl-fas" class="quarto-xref">Table&nbsp;3</a>). For example, this phenotype almost always results in death, either <em>in utero</em> or shortly after birth. Not only did GPT-4 correctly provide the annotation death as ‘always’, when asked whether FAS causes sensory impairments it provided the response ‘always’ with the justification ‘Fetal akinesia sequence typically results in severe sensory impairment due to neurodevelopmental disruption.’ Neurodevelopmental disruption is indeed a hallmark component of FAS (e.g.&nbsp;hydrocephalus, cerebellar hypoplasia) that causes severe impairments across multiple sensory systems <span class="citation" data-cites="chenPrenatalDiagnosisGenetic2012">(<a href="#ref-chenPrenatalDiagnosisGenetic2012" role="doc-biblioref">Chen 2012</a>)</span>. This demonstrates that GPT-4 was able to recover the correct chain of causality from phenotype to clinical outcome.</p>
<p>Our findings highlight the potential of this next generation of natural language processing technologies in significantly contributing to the automation and refinement of data curation in biomedical research. These results have a large number of useful real-world applications, such as prioritising gene therapy candidates <span class="citation" data-cites="murphyIdentificationCellTypespecific2023">(<a href="#ref-murphyIdentificationCellTypespecific2023" role="doc-biblioref">Murphy et al. 2023</a>)</span> and guiding clinical decision-making in rare diseases. As the HPO continues to grow over time, this approach can easily be repeated to keep pace with the rapidly evolving biomedical landscape. Furthermore, this approach can be extended to produce different sets of annotations or be used with any other ontology. Additional use cases include gathering data on the prevalence of each phenotype to approximate their social and financial costs.</p>
<p>One key limitation of our study is the fact that we did not explicitly interrogate GPT-4 to assess how the availability of treatments affected the annotations it produced. For example, there are some very severe conditions for which highly effective treatments and early detection screens are widely available (e.g.&nbsp;syphilis, some forms of melanoma), thus rendering them fully treatable or even curable provided access to modern healthcare. It would therefore be useful to further interrogate GPT-4 to uncover how the availability of treatments influences its responses. Many of our findings here seem to indicate that GPT-4 does take into account quality of care to the extent that health services increase the likelihood of desired outcomes. For example, many of the cancer phenotypes are justified as always or often causing death unless detected and treated early in the disease course. On the other hand, some cancers are justified as rarely causing death if appropriate treatment is provided, which may not always be the case for individuals or populations with access to less access to quality healthcare services. Future efforts could more explicitly ask GPT-4 whether the phenotype would cause death with no or suboptimal treatment.</p>
<p>In addition to these technical challenges, there are multiple factors that need to be considered when trying to prioritise phenotypes for their suitability for gene therapy development. First, while we have attempted to formalise severity here, this is an inherently subjective concept that may vary considerably across different individuals and contexts. For instance, one could ask whether a condition that always causes death is worse than a condition that causes a lifetime of severe disability (e.g.&nbsp;paralysis, blindness, intellectual disability). Metrics such as quality-adjusted life years (QALYs) have been proposed in the past to address these dilemmas by defining health as a function of both the length and quality of life <span class="citation" data-cites="prietoProblemsSolutionsCalculating2003">(<a href="#ref-prietoProblemsSolutionsCalculating2003" role="doc-biblioref">Prieto and Sacristán 2003</a>)</span>. With regards to the financial burden of diseases, in some situations phenotypes which require many years of expensive medical care may be prioritised over those that result in extremely early onset lethality and little opportunity for therapeutic intervention. Another factor that affects the viability of a therapeutic program is the speed, cost and other practical considerations of a clinical trial. For instance, measuring risk of ageing-related respiratory failure over a ten-year period may be impractical in some cases. However, testing for total reversal of an existing severe phenotype could potentially yield faster and more immediately impactful results. If performed in close collaboration with medical ethicists, governmental organisations, advocacy groups and patient families, such cost/benefit assessments could be aided by LLMs through the scalable gathering of relevant data. As AI capabilities continue to advance, the range of applications in which they can be used effectively will continue to grow.</p>
<p>While our study demonstrates the feasibility and utility of AI-driven phenotypic annotation, several limitations must be acknowledged. The reliance on computational algorithms may introduce biases or inaccuracies inherent to the training data, necessitating ongoing validation and refinement of our approach. Additionally, our severity scoring system, while comprehensive, may not capture the full spectrum of phenotypic variability or account for complex gene-environment interactions. Future research should focus on further optimising AI-driven annotation methodologies, incorporating additional data modalities such as genomic and clinical data to enhance accuracy.</p>
<p>In conclusion, our study represents a significant step towards harnessing the power of AI to advance phenotypic annotation and severity assessment across all rare diseases. This resource aims to provide researchers and clinicians with actionable insights that can inform rare disease research and improve the lives of individuals affected by rare diseases.</p>
</section>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<section id="annotating-the-hpo-using-openai-gpt-4" class="level3">
<h3 class="anchored" data-anchor-id="annotating-the-hpo-using-openai-gpt-4">Annotating the HPO using OpenAI GPT-4</h3>
<p>We wrote a Python script to iteratively query GPT-4 via the OpenAI application programming interface (API). The ultimately yielded consistently formatted annotations for 17502 terms within the HPO. Our annotation framework was developed based on previously defined criteria for classifying disease severity <span class="citation" data-cites="Lazarin2014-rz">(<a href="#ref-Lazarin2014-rz" role="doc-biblioref">Lazarin et al. 2014</a>)</span>. We sought to evaluate whether each phenotype directly caused a given severity-related clinical characteristic, including: intellectual disability, death, impaired mobility, physical malformations, blindness, sensory impairments, immunodeficiency, cancer, reduced fertility, and/or had a congenital onset. Through prompt engineering we found that the performance of GPT-4 improved when we incorporated a scale associated with each effect and required a justification for each response. We asked how frequently the given phenotype directly causes each clinical characteristic - whether it never, rarely, often, or always occurred. Furthermore, our prompt design revealed that the optimal trade-off between the number of phenotypes and performance (in terms of producing the desired annotations, and adhering to the formatting requirements) was achieved when inputting no more than two or three phenotypes per prompt. An example prompt can be seen in <a href="#fig-occurrence" class="quarto-xref">Figure&nbsp;1</a>. Only 2 phenotypes were included per prompt in order to 1) avoid exceeding per-query token limits, and 2) prevent the breakdown of GPT-4 performance due to long-form text input, which is presently a known limitation common to many LLMs including GPT-4 <span class="citation" data-cites="weiLongformFactualityLarge2024">(<a href="#ref-weiLongformFactualityLarge2024" role="doc-biblioref">Wei et al. 2024</a>)</span>.</p>
</section>
<section id="calculating-the-true-positive-rate" class="level3">
<h3 class="anchored" data-anchor-id="calculating-the-true-positive-rate">Calculating the true positive rate</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [20]:</pre></div><div class="cell" data-tbl-colwidths="[40,40,20]">

<div class="cell-container"><div class="cell-decorator"><pre>In [21]:</pre></div><div><div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>queries <span class="ot">&lt;-</span> <span class="fu">eval</span>(<span class="fu">formals</span>(HPOExplorer<span class="sc">::</span>search_hpo)<span class="sc">$</span>queries)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>query_hits_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Clinical characteristic"</span><span class="ot">=</span>stringr<span class="sc">::</span><span class="fu">str_to_sentence</span>(<span class="fu">gsub</span>(<span class="st">"_"</span>,<span class="st">" "</span>,<span class="fu">names</span>(queries))),</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"HPO queries"</span><span class="ot">=</span><span class="fu">trimws</span>(<span class="fu">gsub</span>(<span class="st">"^"</span>,<span class="st">""</span>,</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>                            <span class="fu">sapply</span>(queries, <span class="cf">function</span>(x){<span class="fu">paste</span>(<span class="fu">shQuote</span>(x),<span class="at">collapse=</span><span class="st">"; "</span>)}), </span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">fixed =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"True positive HPO IDs"</span><span class="ot">=</span><span class="fu">sapply</span>(query_hits[<span class="fu">names</span>(queries)], data.table<span class="sc">::</span>uniqueN)</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">as.data.frame</span>(query_hits_dt))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-true-positives" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-true-positives-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: For each clinical characteristic, any terms that matched
</figcaption>
<div aria-describedby="tbl-true-positives-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="cell table table-sm table-striped small">
<colgroup>
<col style="width: 40%">
<col style="width: 40%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Clinical characteristic</th>
<th style="text-align: left;">HPO queries</th>
<th style="text-align: right;">True positive HPO IDs</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Intellectual disability</td>
<td style="text-align: left;">‘Intellectual disability’; ‘Mental deterioration’</td>
<td style="text-align: right;">19</td>
</tr>
<tr class="even">
<td style="text-align: left;">Impaired mobility</td>
<td style="text-align: left;">‘Gait disturbance’; ‘Diminished movement’; ’ mobility’</td>
<td style="text-align: right;">319</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Physical malformations</td>
<td style="text-align: left;">‘malformation’</td>
<td style="text-align: right;">78</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blindness</td>
<td style="text-align: left;">‘blindness’</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sensory impairments</td>
<td style="text-align: left;">‘Abnormality of vision’; ‘Abnormality of the sense of smell’; ‘Abnormality of taste sensation’; ‘Somatic sensory dysfunction’; ‘Hearing abnormality’</td>
<td style="text-align: right;">252</td>
</tr>
<tr class="even">
<td style="text-align: left;">Immunodeficiency</td>
<td style="text-align: left;">‘Immunodeficiency’; ‘Impaired antigen-specific response’</td>
<td style="text-align: right;">29</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Cancer</td>
<td style="text-align: left;">‘Cancer’; ‘malignant’; ‘carcinoma’</td>
<td style="text-align: right;">56</td>
</tr>
<tr class="even">
<td style="text-align: left;">Reduced fertility</td>
<td style="text-align: left;">‘Decreased fertility’; ‘Hypogonadism’</td>
<td style="text-align: right;">9</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div></div></div>
</div></div>
<p>A true positive rate was calculated as a measure of the recall of the GPT-4 annotations. This was achieved by identifying specific branches within the HPO that would contain phenotypes that would reliably indicate the occurrence of certain clinical characteristics, and using all descendants of this HPO branch as true positives. For example, all descendants of the terms ‘Intellectual disability’ (HP:0001249) or ‘Mental deterioration’ (HP:0001268) should be annotated as always or often causing intellectual disability (<a href="#tbl-true-positives" class="quarto-xref">Table&nbsp;1</a>).</p>
</section>
<section id="quantifying-phenotypic-severity-1" class="level3">
<h3 class="anchored" data-anchor-id="quantifying-phenotypic-severity-1">Quantifying phenotypic severity</h3>
<p>The GPT-4 generated annotation occurrences were converted into a semi-quantitative scoring system, with ‘always’ corresponding to 3, ‘often’ to 2, ‘rarely’ to 1, and ‘never’ to 0. These scores were then weighted by a severity metric on a scale of 1-5, with 5 representing the highest severity, as determined by the provided annotations (<a href="#tbl-metric-weights" class="quarto-xref">Table&nbsp;2</a>). Subsequently, the weighted scores underwent normalisation to yield a final quantitative severity score ranging from 0-100, with 100 signifying the maximum severity score attainable.</p>
<p>Let us denote:</p>
<ul>
<li><p><span class="math inline">\(p\)</span> : a phenotype in the HPO.</p></li>
<li><p><span class="math inline">\(j\)</span> : the identity of a given annotation metric (i.e.&nbsp;clinical characteristic, such as ‘intellectual disability’ or ‘congenital onset’).</p></li>
<li><p><span class="math inline">\(W_j\)</span>: the assigned weight of metric <span class="math inline">\(j\)</span>.</p></li>
<li><p><span class="math inline">\(F_j\)</span>: the maximum possible value for metric <span class="math inline">\(j\)</span> (equivalent across all <span class="math inline">\(j\)</span>).</p></li>
<li><p><span class="math inline">\(F_{pj}\)</span> : the numerically encoded value of annotation metric <span class="math inline">\(j\)</span> for phenotype <span class="math inline">\(p\)</span>.</p></li>
<li><p><span class="math inline">\(NSS_p\)</span>: the final composite severity score for phenotype <span class="math inline">\(p\)</span> after applying normalisation to align values to a 0-100 scale and ensure equivalent meaning regardless of which other phenotypes are being analysed in addition to <span class="math inline">\(p\)</span>. This allows for direct comparability of severity scores across studies with different sets of phenotypes.</p></li>
</ul>
<p><a href="figures/eq5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="figures/eq5.png" height="300"></a></p>
<p>Computing normalised severity score from encoded GPT-4 annotations.</p>
<p><br>
<br>
</p>
<p><br>
</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [22]:</pre></div><div class="cell">

<div class="cell-container"><div class="cell-decorator"><pre>In [23]:</pre></div><div><div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Get weights from HPOExplorer::gpt_annot_codify function defaults</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">formals</span>(HPOExplorer<span class="sc">::</span>gpt_annot_codify)</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>code_dict <span class="ot">&lt;-</span> <span class="fu">eval</span>(args<span class="sc">$</span>code_dict)<span class="sc">|&gt;</span><span class="fu">sort</span>(<span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(code_dict) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(stringr<span class="sc">::</span><span class="fu">str_to_sentence</span>(<span class="fu">names</span>(code_dict)),</span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>                           <span class="st">" ("</span>,code_dict,<span class="st">")"</span>)</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>weights_dict <span class="ot">&lt;-</span> <span class="fu">eval</span>(args<span class="sc">$</span>weights_dict)<span class="sc">|&gt;</span><span class="fu">unlist</span>()<span class="sc">|&gt;</span><span class="fu">sort</span>(<span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(weights_dict) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(stringr<span class="sc">::</span><span class="fu">str_to_sentence</span>(<span class="fu">gsub</span>(<span class="st">"_"</span>,<span class="st">" "</span>,<span class="fu">names</span>(weights_dict))),</span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a>                           <span class="st">" ("</span>,weights_dict,<span class="st">")"</span>)</span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Create the outer product of the two weight vectors</span></span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a>tbl_metric_weights <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(<span class="fu">outer</span>(weights_dict,code_dict, <span class="st">"*"</span>),</span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">keep.rownames =</span> <span class="st">"Clinical characteristic"</span>)</span>
<span id="cb108-12"><a href="#cb108-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-13"><a href="#cb108-13" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">as.data.frame</span>(tbl_metric_weights))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-metric-weights" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-metric-weights-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2: Weighted scores for each clinical characteristic and GPT-4 response category.
</figcaption>
<div aria-describedby="tbl-metric-weights-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="do-not-create-environment cell table table-sm table-striped small">
<colgroup>
<col style="width: 40%">
<col style="width: 15%">
<col style="width: 14%">
<col style="width: 15%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Clinical characteristic</th>
<th style="text-align: right;">Always (3)</th>
<th style="text-align: right;">Often (2)</th>
<th style="text-align: right;">Rarely (1)</th>
<th style="text-align: right;">Never (0)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Death (6)</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Intellectual disability (5)</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Impaired mobility (4)</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blindness (4)</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Physical malformations (3)</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sensory impairments (3)</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Immunodeficiency (3)</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cancer (3)</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Reduced fertility (1)</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Congenital onset (1)</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div></div></div>
</div></div>
</section>
<section id="severity-classes-1" class="level3">
<h3 class="anchored" data-anchor-id="severity-classes-1">Severity classes</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [24]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>tiers_dict <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">eval</span>(<span class="fu">formals</span>(HPOExplorer<span class="sc">::</span>gpt_annot_class)<span class="sc">$</span>tiers_dict))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>The decision tree algorithm used in <span class="citation" data-cites="Lazarin2014-rz">Lazarin et al. (<a href="#ref-Lazarin2014-rz" role="doc-biblioref">2014</a>)</span> was adapted here for use with the GPT-4 annotations. This algorithm first assigned each annotation metric to a tier, where Tier 1 indicated the most severe annotation metrics and Tier 4 indicated the least severe annotation metrics (‘death’=1, ‘intellectual disability’=1, ‘impaired mobility’=2, ‘physical malformations’=2, ‘blindness’=3, ‘sensory impairments’=3, ‘immunodeficiency’=3, ‘cancer’=3, ‘reduced fertility’=4). If a phenotype often or always caused more than one Tier 1 annotation, it was assigned a severity class of “Profound”. If the phenotype often or always caused only one Tier 1 annotation, it was assigned a severity class of “Severe”. A “Severe” class assignment was also assigned if the phenotype often or always caused three or more Tier 2 and Tier3 annotations. If the phenotype often or always caused at least one Tier 2 annotation, it was assigned a severity class of “Moderate”. All remaining phenotypes were was assigned a severity class of “Mild”. In cases where the phenotype mapped to more than one class, only the most severe class was used. This procedure is implemented within the function <code>HPOExplorer::gpt_annot_class</code>.</p>
</section>
<section id="correlations-between-severity-metrics-1" class="level3">
<h3 class="anchored" data-anchor-id="correlations-between-severity-metrics-1">Correlations between severity metrics</h3>
<p>To assess the correlation structure between each severity metric, as well as between the composite severity score and each metric, we computed Pearson correlation coefficients for all pairwise combinations of these variables using the numerically encoded metric values. The correlation matrix was visualised using a heatmap, with the colour intensity representing the strength of the correlation (<a href="#fig-metric-corplot" class="quarto-xref">Figure&nbsp;8</a>).</p>
</section>
</section>
<section id="data-and-code-availability-statement" class="level2">
<h2 class="anchored" data-anchor-id="data-and-code-availability-statement">Data and code availability statement</h2>
<p>All code and data used in this study are available on GitHub at:</p>
<p><a href="https://github.com/neurogenomics/gpt_hpo_annotations" class="uri">https://github.com/neurogenomics/gpt_hpo_annotations</a></p>
<p>The GPT-4 annotations for all HPO phenotypes are made available through the R function <code>HPOExplorer::gpt_annot_read</code> or in CSV format at:</p>
<p><a href="https://github.com/neurogenomics/gpt_hpo_annotations/tree/master/data" class="uri">https://github.com/neurogenomics/gpt_hpo_annotations/tree/master/data</a></p>
<p>A fully reproducible version of this Quarto manuscript can be found at:</p>
<p><a href="https://github.com/neurogenomics/gpt_hpo_annotations/blob/master/manuscript.qmd" class="uri">https://github.com/neurogenomics/gpt_hpo_annotations/blob/master/manuscript.qmd</a></p>
</section>
<section id="acknowledgements" class="level2">
<h2 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h2>
<p>We would like to thank members of the Monarch Initiative for their insight and feedback throughout this project. In particular, Peter Robinson.</p>
<section id="funding" class="level3">
<h3 class="anchored" data-anchor-id="funding">Funding</h3>
<p>This work was supported by a UK Dementia Research Institute (UK DRI) Future Leaders Fellowship [MR/T04327X/1] and the UK DRI which receives its funding from UK DRI Ltd, funded by the UK Medical Research Council, Alzheimer’s Society and Alzheimer’s Research UK.</p>
</section>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-boltonBioMedLM7BParameter2024" class="csl-entry" role="listitem">
Bolton, Elliot, Abhinav Venigalla, Michihiro Yasunaga, David Hall, Betty Xiong, Tony Lee, Roxana Daneshjou, et al. 2024. <span>“BioMedLM: A 2.7B Parameter Language Model Trained on Biomedical Text.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2403.18421">https://doi.org/10.48550/arXiv.2403.18421</a>.
</div>
<div id="ref-caufieldStructuredPromptInterrogation2023" class="csl-entry" role="listitem">
Caufield, J. Harry, Harshad Hegde, Vincent Emonet, Nomi L. Harris, Marcin P. Joachimiak, Nicolas Matentzoglu, HyeongSik Kim, et al. 2023. <span>“Structured Prompt Interrogation and Recursive Extraction of Semantics (SPIRES): A Method for Populating Knowledge Bases Using Zero-Shot Learning.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2304.02711">https://doi.org/10.48550/arXiv.2304.02711</a>.
</div>
<div id="ref-chenPrenatalDiagnosisGenetic2012" class="csl-entry" role="listitem">
Chen, Chih-Ping. 2012. <span>“Prenatal Diagnosis and Genetic Analysis of Fetal Akinesia Deformation Sequence and Multiple Pterygium Syndrome Associated with Neuromuscular Junction Disorders: A Review.”</span> <em>Taiwanese Journal of Obstetrics and Gynecology</em> 51 (1): 12–17. <a href="https://doi.org/10.1016/j.tjog.2012.01.004">https://doi.org/10.1016/j.tjog.2012.01.004</a>.
</div>
<div id="ref-chengExploringPotentialGPT42023" class="csl-entry" role="listitem">
Cheng, Kunming, Qiang Guo, Yongbin He, Yanqiu Lu, Shuqin Gu, and Haiyang Wu. 2023. <span>“Exploring the Potential of GPT-4 in Biomedical Engineering: The Dawn of a New Era.”</span> <em>Annals of Biomedical Engineering</em> 51 (8): 1645–53. <a href="https://doi.org/10.1007/s10439-023-03221-1">https://doi.org/10.1007/s10439-023-03221-1</a>.
</div>
<div id="ref-garagnaniSyndromesAssociatedSyndactyly2013" class="csl-entry" role="listitem">
Garagnani, Lorenzo, and Gillian D. Smith. 2013. <span>“Syndromes Associated with Syndactyly.”</span> In <em>The Pediatric Upper Extremity</em>, edited by Joshua M. Abzug, Scott Kozin, and Dan A. Zlotolow, 1–31. Springer. <a href="https://doi.org/10.1007/978-1-4614-8758-6_14-1">https://doi.org/10.1007/978-1-4614-8758-6_14-1</a>.
</div>
<div id="ref-garganoHumanPhenotypeOntology2024" class="csl-entry" role="listitem">
Gargano, Michael A, Nicolas Matentzoglu, Ben Coleman, Eunice B Addo-Lartey, Anna V Anagnostopoulos, Joel Anderton, Paul Avillach, et al. 2024. <span>“The Human Phenotype Ontology in 2024: Phenotypes Around the World.”</span> <em>Nucleic Acids Research</em> 52 (D1): D1333–46. <a href="https://doi.org/10.1093/nar/gkad1005">https://doi.org/10.1093/nar/gkad1005</a>.
</div>
<div id="ref-gleesonMolarToothSign2004" class="csl-entry" role="listitem">
Gleeson, Joseph G., Lesley C. Keeler, Melissa A. Parisi, Sarah E. Marsh, Phillip F. Chance, Ian A. Glass, John M. Graham Jr, Bernard L. Maria, A. James Barkovich, and William B. Dobyns. 2004. <span>“Molar Tooth Sign of the Midbrain–Hindbrain Junction: Occurrence in Multiple Distinct Syndromes.”</span> <em>American Journal of Medical Genetics Part A</em> 125A (2): 125–34. <a href="https://doi.org/10.1002/ajmg.a.20437">https://doi.org/10.1002/ajmg.a.20437</a>.
</div>
<div id="ref-guDomainSpecificLanguageModel2021" class="csl-entry" role="listitem">
Gu, Yu, Robert Tinn, Hao Cheng, Michael Lucas, Naoto Usuyama, Xiaodong Liu, Tristan Naumann, Jianfeng Gao, and Hoifung Poon. 2021. <span>“Domain-Specific Language Model Pretraining for Biomedical Natural Language Processing.”</span> <em>ACM Transactions on Computing for Healthcare</em> 3 (1): 2:1–23. <a href="https://doi.org/10.1145/3458754">https://doi.org/10.1145/3458754</a>.
</div>
<div id="ref-janikAspectsHumanMemory2024" class="csl-entry" role="listitem">
Janik, Romuald A. 2024. <span>“Aspects of Human Memory and Large Language Models.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2311.03839">https://doi.org/10.48550/arXiv.2311.03839</a>.
</div>
<div id="ref-jonesStatisticalInterpretationTerm1972" class="csl-entry" role="listitem">
Jones, Karen Spärck. 1972. <span>“A Statistical Interpretation of Term Specificity and Its Application in Retrieval.”</span> <em>Journal of Documentation</em> 28 (1): 11–21. <a href="https://doi.org/10.1108/eb026526">https://doi.org/10.1108/eb026526</a>.
</div>
<div id="ref-kohlerHumanPhenotypeOntology2021" class="csl-entry" role="listitem">
Köhler, Sebastian, Michael Gargano, Nicolas Matentzoglu, Leigh C Carmody, David Lewis-Smith, Nicole A Vasilevsky, Daniel Danis, et al. 2021. <span>“The Human Phenotype Ontology in 2021.”</span> <em>Nucleic Acids Research</em> 49 (D1): D1207–17. <a href="https://doi.org/10.1093/nar/gkaa1043">https://doi.org/10.1093/nar/gkaa1043</a>.
</div>
<div id="ref-labrakBioMistralCollectionOpenSource2024" class="csl-entry" role="listitem">
Labrak, Yanis, Adrien Bazoge, Emmanuel Morin, Pierre-Antoine Gourraud, Mickael Rouvier, and Richard Dufour. 2024. <span>“BioMistral: A Collection of Open-Source Pretrained Large Language Models for Medical Domains.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2402.10373">https://doi.org/10.48550/arXiv.2402.10373</a>.
</div>
<div id="ref-Lazarin2014-rz" class="csl-entry" role="listitem">
Lazarin, Gabriel A, Felicia Hawthorne, Nicholas S Collins, Elizabeth A Platt, Eric A Evans, and Imran S Haque. 2014. <span>“Systematic Classification of Disease Severity for Evaluation of Expanded Carrier Screening Panels.”</span> <em>PLoS One</em> 9 (12): e114391.
</div>
<div id="ref-luoBioGPTGenerativePretrained2022" class="csl-entry" role="listitem">
Luo, Renqian, Liai Sun, Yingce Xia, Tao Qin, Sheng Zhang, Hoifung Poon, and Tie-Yan Liu. 2022. <span>“BioGPT: Generative Pre-Trained Transformer for Biomedical Text Generation and Mining.”</span> <em>Briefings in Bioinformatics</em> 23 (6): bbac409. <a href="https://doi.org/10.1093/bib/bbac409">https://doi.org/10.1093/bib/bbac409</a>.
</div>
<div id="ref-marcinkevicsInterpretabilityExplainabilityMachine2023" class="csl-entry" role="listitem">
Marcinkevičs, Ričards, and Julia E. Vogt. 2023. <span>“Interpretability and Explainability: A Machine Learning Zoo Mini-Tour.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2012.01805">https://doi.org/10.48550/arXiv.2012.01805</a>.
</div>
<div id="ref-mcduffAccurateDifferentialDiagnosis2023" class="csl-entry" role="listitem">
McDuff, Daniel, Mike Schaekermann, Tao Tu, Anil Palepu, Amy Wang, Jake Garrison, Karan Singhal, et al. 2023. <span>“Towards Accurate Differential Diagnosis with Large Language Models.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2312.00164">https://doi.org/10.48550/arXiv.2312.00164</a>.
</div>
<div id="ref-mikolovEfficientEstimationWord2013" class="csl-entry" role="listitem">
Mikolov, Tomas, Kai Chen, Greg Corrado, and Jeffrey Dean. 2013. <span>“Efficient Estimation of Word Representations in Vector Space.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.1301.3781">https://doi.org/10.48550/arXiv.1301.3781</a>.
</div>
<div id="ref-mungallMonarchInitiativeIntegrative2017a" class="csl-entry" role="listitem">
Mungall, Christopher J., Julie A. McMurry, Sebastian Köhler, James P. Balhoff, Charles Borromeo, Matthew Brush, Seth Carbon, et al. 2017. <span>“The Monarch Initiative: An Integrative Data and Analytic Platform Connecting Phenotypes to Genotypes Across Species.”</span> <em>Nucleic Acids Research</em> 45 (D1): D712–22. <a href="https://doi.org/10.1093/nar/gkw1128">https://doi.org/10.1093/nar/gkw1128</a>.
</div>
<div id="ref-murphyIdentificationCellTypespecific2023" class="csl-entry" role="listitem">
Murphy, Kitty B., Robert Gordon-Smith, Jai Chapman, Momoko Otani, Brian M. Schilder, and Nathan G. Skene. 2023. <span>“Identification of Cell Type-Specific Gene Targets Underlying Thousands of Rare Diseases and Subtraits.”</span> <em>medRxiv</em>. <a href="https://doi.org/10.1101/2023.02.13.23285820">https://doi.org/10.1101/2023.02.13.23285820</a>.
</div>
<div id="ref-noriCanGeneralistFoundation2023" class="csl-entry" role="listitem">
Nori, Harsha, Yin Tat Lee, Sheng Zhang, Dean Carignan, Richard Edgar, Nicolo Fusi, Nicholas King, et al. 2023. <span>“Can Generalist Foundation Models Outcompete Special-Purpose Tuning? Case Study in Medicine.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2311.16452">https://doi.org/10.48550/arXiv.2311.16452</a>.
</div>
<div id="ref-oneilPhenomicsAssistantInterface2024" class="csl-entry" role="listitem">
O’Neil, Shawn T., Kevin Schaper, Glass Elsarboukh, Justin T. Reese, Sierra A. T. Moxon, Nomi L. Harris, Monica C. Munoz-Torres, Peter N. Robinson, Melissa A. Haendel, and Christopher J. Mungall. 2024. <span>“Phenomics Assistant: An Interface for LLM-Based Biomedical Knowledge Graph Exploration.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2024.01.31.578275">https://doi.org/10.1101/2024.01.31.578275</a>.
</div>
<div id="ref-ochoaOpenTargetsPlatform2021" class="csl-entry" role="listitem">
Ochoa, David, Andrew Hercules, Miguel Carmona, Daniel Suveges, Asier Gonzalez-Uriarte, Cinzia Malangone, Alfredo Miranda, et al. 2021. <span>“Open Targets Platform: Supporting Systematic Drug–Target Identification and Prioritisation.”</span> <em>Nucleic Acids Research</em> 49 (D1): D1302–10. <a href="https://doi.org/10.1093/nar/gkaa1027">https://doi.org/10.1093/nar/gkaa1027</a>.
</div>
<div id="ref-openaiGPT4TechnicalReport2024" class="csl-entry" role="listitem">
OpenAI, Josh Achiam, Steven Adler, Sandhini Agarwal, Lama Ahmad, Ilge Akkaya, Florencia Leoni Aleman, et al. 2024. <span>“GPT-4 Technical Report.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2303.08774">https://doi.org/10.48550/arXiv.2303.08774</a>.
</div>
<div id="ref-panLargeLanguageModels2023" class="csl-entry" role="listitem">
Pan, Jeff Z., Simon Razniewski, Jan-Christoph Kalo, Sneha Singhania, Jiaoyan Chen, Stefan Dietze, Hajira Jabeen, et al. 2023. <span>“Large Language Models and Knowledge Graphs: Opportunities and Challenges.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2308.06374">https://doi.org/10.48550/arXiv.2308.06374</a>.
</div>
<div id="ref-prietoProblemsSolutionsCalculating2003" class="csl-entry" role="listitem">
Prieto, Luis, and José A Sacristán. 2003. <span>“Problems and Solutions in Calculating Quality-Adjusted Life Years (QALYs).”</span> <em>Health and Quality of Life Outcomes</em> 1 (December): 80. <a href="https://doi.org/10.1186/1477-7525-1-80">https://doi.org/10.1186/1477-7525-1-80</a>.
</div>
<div id="ref-putmanMonarchInitiative20242024" class="csl-entry" role="listitem">
Putman, Tim E, Kevin Schaper, Nicolas Matentzoglu, Vincent P Rubinetti, Faisal S Alquaddoomi, Corey Cox, J Harry Caufield, et al. 2024. <span>“The Monarch Initiative in 2024: An Analytic Platform Integrating Phenotypes, Genes&nbsp;and Diseases Across Species.”</span> <em>Nucleic Acids Research</em> 52 (D1): D938–49. <a href="https://doi.org/10.1093/nar/gkad1082">https://doi.org/10.1093/nar/gkad1082</a>.
</div>
<div id="ref-shinBioMegatronLargerBiomedical2020" class="csl-entry" role="listitem">
Shin, Hoo-Chang, Yang Zhang, Evelina Bakhturina, Raul Puri, Mostofa Patwary, Mohammad Shoeybi, and Raghav Mani. 2020. <span>“BioMegatron: Larger Biomedical Domain Language Model.”</span> In <em>Proceedings of the 2020 Conference on Empirical Methods in Natural Language Processing (EMNLP)</em>, edited by Bonnie Webber, Trevor Cohn, Yulan He, and Yang Liu, 4700–4706. Association for Computational Linguistics. <a href="https://doi.org/10.18653/v1/2020.emnlp-main.379">https://doi.org/10.18653/v1/2020.emnlp-main.379</a>.
</div>
<div id="ref-singhalLargeLanguageModels2023" class="csl-entry" role="listitem">
Singhal, Karan, Shekoofeh Azizi, Tao Tu, S. Sara Mahdavi, Jason Wei, Hyung Won Chung, Nathan Scales, et al. 2023. <span>“Large Language Models Encode Clinical Knowledge.”</span> <em>Nature</em>, July, 1–9. <a href="https://doi.org/10.1038/s41586-023-06291-2">https://doi.org/10.1038/s41586-023-06291-2</a>.
</div>
<div id="ref-singhalExpertLevelMedicalQuestion2023" class="csl-entry" role="listitem">
Singhal, Karan, Tao Tu, Juraj Gottweis, Rory Sayres, Ellery Wulczyn, Le Hou, Kevin Clark, et al. 2023. <span>“Towards Expert-Level Medical Question Answering with Large Language Models.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2305.09617">https://doi.org/10.48550/arXiv.2305.09617</a>.
</div>
<div id="ref-toroDynamicRetrievalAugmented2023" class="csl-entry" role="listitem">
Toro, Sabrina, Anna V. Anagnostopoulos, Sue Bello, Kai Blumberg, Rhiannon Cameron, Leigh Carmody, Alexander D. Diehl, et al. 2023. <span>“Dynamic Retrieval Augmented Generation of Ontologies Using Artificial Intelligence (DRAGON-AI).”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2312.10904">https://doi.org/10.48550/arXiv.2312.10904</a>.
</div>
<div id="ref-vanveenAdaptedLargeLanguage2024" class="csl-entry" role="listitem">
Van Veen, Dave, Cara Van Uden, Louis Blankemeier, Jean-Benoit Delbrouck, Asad Aali, Christian Bluethgen, Anuj Pareek, et al. 2024. <span>“Adapted Large Language Models Can Outperform Medical Experts in Clinical Text Summarization.”</span> <em>Nature Medicine</em>, February, 1–9. <a href="https://doi.org/10.1038/s41591-024-02855-5">https://doi.org/10.1038/s41591-024-02855-5</a>.
</div>
<div id="ref-weiLongformFactualityLarge2024" class="csl-entry" role="listitem">
Wei, Jerry, Chengrun Yang, Xinying Song, Yifeng Lu, Nathan Hu, Jie Huang, Dustin Tran, et al. 2024. <span>“Long-Form Factuality in Large Language Models.”</span> <em>arXiv</em>. <a href="http://arxiv.org/abs/2403.18802">http://arxiv.org/abs/2403.18802</a>.
</div>
<div id="ref-zhangBiomedGPTUnifiedGeneralist2023" class="csl-entry" role="listitem">
Zhang, Kai, Jun Yu, Zhiling Yan, Yixin Liu, Eashan Adhikarla, Sunyang Fu, Xun Chen, et al. 2023. <span>“BiomedGPT: A Unified and Generalist Biomedical Generative Pre-Trained Transformer for Vision, Language, and Multimodal Tasks.”</span> <em>arXiv</em>. <a href="https://doi.org/10.48550/arXiv.2305.17100">https://doi.org/10.48550/arXiv.2305.17100</a>.
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="supplementary-materials" class="level2">
<h2 class="anchored" data-anchor-id="supplementary-materials">Supplementary Materials</h2>
<section id="supplementary-figures" class="level3">
<h3 class="anchored" data-anchor-id="supplementary-figures">Supplementary Figures</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [25]:</pre></div><div id="cell-fig-consist-vs-ontLvl" class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>consist_vs_ontLvl<span class="sc">$</span>labels<span class="sc">$</span>caption <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>consist_vs_ontLvl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-consist-vs-ontLvl" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-consist-vs-ontLvl-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-consist-vs-ontLvl-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6" data-glightbox="description: .lightbox-desc-6"><img src="manuscript_files/figure-html/fig-consist-vs-ontLvl-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-consist-vs-ontLvl-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Relationship between the consistency of GPT-4 annotations (using the stringent criterion) and the level of each phenotype within the HPO ontology (with the number of phenotypes in parentheses). Greater ontology levels (x-axis) indicate more specific phenotypes. The subtitle indicates summary statistics for the overall relationship between HPO level and the proportion of phenotypes that were annotated consistently. The p-values above each bar indicate whether the distribution of consistent/inconsistent annotations, within a given HPO level, significantly deviate from the expected null distribution.
</figcaption>
</figure>
</div>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [26]:</pre></div><div id="cell-fig-severity-histo" class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>gpt_annot_plot_out<span class="sc">$</span>gp3 <span class="sc">+</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Severity score"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-severity-histo" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-severity-histo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-severity-histo-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7" data-glightbox="description: .lightbox-desc-7"><img src="manuscript_files/figure-html/fig-severity-histo-1.png" class="img-fluid figure-img" width="960"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-severity-histo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Distribution of the composite GPT-4 severity score of the severity scores for all HPO terms.
</figcaption>
</figure>
</div>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [27]:</pre></div><div id="cell-fig-severity-boxplot" class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>gpt_annot_plot_out<span class="sc">$</span>gp2 <span class="sc">+</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"GnBu"</span>, <span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Severity score"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-severity-boxplot" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-severity-boxplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-severity-boxplot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" data-glightbox="description: .lightbox-desc-8"><img src="manuscript_files/figure-html/fig-severity-boxplot-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-severity-boxplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Boxplot showing the relationship between composite severity score (y-axis) and the frequency response categories within each annotation type.
</figcaption>
</figure>
</div>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [28]:</pre></div><div id="cell-fig-metric-corplot" class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(cor_metrics<span class="sc">$</span>P) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>corrplot<span class="sc">::</span><span class="fu">corrplot</span>(cor_metrics<span class="sc">$</span>r,</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">method =</span> <span class="fu">c</span>(<span class="st">"circle"</span>, <span class="st">"square"</span>, <span class="st">"ellipse"</span>, <span class="st">"number"</span>, <span class="st">"shade"</span>, <span class="st">"color"</span>, <span class="st">"pie"</span>)[<span class="dv">2</span>],</span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">tl.srt =</span> <span class="dv">45</span>,</span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">tl.col =</span> <span class="st">"grey20"</span>,</span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># insig='blank',</span></span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">addCoef.col =</span><span class="st">'grey20'</span>, <span class="at">number.cex =</span> <span class="fl">0.8</span>,</span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">order=</span><span class="st">"hclust"</span>,</span>
<span id="cb114-9"><a href="#cb114-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">addCoefasPercent=</span><span class="cn">TRUE</span>,</span>
<span id="cb114-10"><a href="#cb114-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">p.mat =</span> cor_metrics<span class="sc">$</span>P)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-metric-corplot" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-metric-corplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-metric-corplot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9" data-glightbox="description: .lightbox-desc-9"><img src="manuscript_files/figure-html/fig-metric-corplot-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-metric-corplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Pearson correlations between each individual severity metric and the composite severity score (‘severity_score_gpt’).
</figcaption>
</figure>
</div>
</div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [29]:</pre></div><div id="cell-fig-severity-class" class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>fig_severity_class</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-severity-class" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-severity-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="manuscript_files/figure-html/fig-severity-class-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10" data-glightbox="description: .lightbox-desc-10"><img src="manuscript_files/figure-html/fig-severity-class-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-severity-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Distribution of the composite GPT-4 severity score introduced in this paper (y-axis) by an approximation of the severity class system introduced <span class="citation" data-cites="Lazarin2014-rz">Lazarin et al. (<a href="#ref-Lazarin2014-rz" role="doc-biblioref">2014</a>)</span> (x-axis). While these are different schemes for ranking phenotype severity, there is a strong correspondence between them (see summary statistics in subtitle). The sample size (number of phenotypes) is shown in parentheses along the x-axis.
</figcaption>
</figure>
</div>
</div>
</div></div>
<div style="page-break-after: always;"></div>
</section>
<section id="supplementary-tables" class="level3">
<h3 class="anchored" data-anchor-id="supplementary-tables">Supplementary Tables</h3>
<div class="cell-container"><div class="cell-decorator"><pre>In [30]:</pre></div><div class="cell" data-tbl-colwidths="[20,20,60]">

<div class="cell-container"><div class="cell-decorator"><pre>In [31]:</pre></div><div><div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>annot_cols <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="fu">grep</span>(<span class="st">"justification"</span>,<span class="fu">names</span>(gpt_annot), </span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">value =</span> <span class="cn">TRUE</span>, <span class="at">invert =</span> <span class="cn">TRUE</span>),</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>                       <span class="fu">c</span>(<span class="st">"hpo_name"</span>,<span class="st">"hpo_id"</span>,<span class="st">"pheno_count"</span>))</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>tbl_fas <span class="ot">&lt;-</span> gpt_annot[hpo_name<span class="sc">==</span><span class="st">'Fetal akinesia sequence'</span>,]<span class="sc">|&gt;</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">melt.data.table</span>(<span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">"hpo_name"</span>,<span class="st">"hpo_id"</span>), </span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">measure.vars =</span> <span class="fu">list</span>(</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>                                <span class="at">Annotation=</span>annot_cols,</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>                                <span class="at">Justification=</span><span class="fu">grep</span>(<span class="st">"justification"</span>,<span class="fu">names</span>(gpt_annot), </span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>                                )</span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>                              )</span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>tbl_fas[,<span class="st">"Clinical characteristic"</span><span class="sc">:</span><span class="er">=</span>stringr<span class="sc">::</span><span class="fu">str_to_sentence</span>(<span class="fu">gsub</span>(<span class="st">"_"</span>,<span class="st">" "</span>,annot_cols))]</span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>(tbl_fas[,<span class="fu">c</span>(<span class="st">"Clinical characteristic"</span>,<span class="st">"Annotation"</span>,<span class="st">"Justification"</span>)])</span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-fas" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-fas-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;3: Severity nnotations generated for GPT-4 annotations for the HPO phenotype ‘Fetal akinesia sequence’ (HP:000198).
</figcaption>
<div aria-describedby="tbl-fas-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="cell table table-sm table-striped small">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 60%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Clinical characteristic</th>
<th style="text-align: left;">Annotation</th>
<th style="text-align: left;">Justification</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Intellectual disability</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence typically results in severe neurodevelopmental impairment, including intellectual disability.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Death</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence is typically fatal in utero or shortly after birth.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Impaired mobility</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence results in severe physical impairment, including impaired mobility.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Physical malformations</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence is associated with multiple physical malformations.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Blindness</td>
<td style="text-align: left;">often</td>
<td style="text-align: left;">Visual impairment is common in surviving individuals with fetal akinesia sequence due to neurodevelopmental impairment.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sensory impairments</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence typically results in severe sensory impairment due to neurodevelopmental disruption.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Immunodeficiency</td>
<td style="text-align: left;">rarely</td>
<td style="text-align: left;">While not a direct feature, some individuals with fetal akinesia sequence may have associated immune abnormalities.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cancer</td>
<td style="text-align: left;">never</td>
<td style="text-align: left;">Fetal akinesia sequence does not cause cancer.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Reduced fertility</td>
<td style="text-align: left;">often</td>
<td style="text-align: left;">Given the severe physical impairments associated with fetal akinesia sequence, fertility is likely to be reduced in surviving individuals.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Congenital onset</td>
<td style="text-align: left;">always</td>
<td style="text-align: left;">Fetal akinesia sequence is a congenital disorder.</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div></div></div>
</div></div>
<div class="hidden" aria-hidden="true">
<span class="glightbox-desc lightbox-desc-1">Figure&nbsp;1: GPT-4 was able to annotate all human phenotypes based on whether they are always/often/rarely/never associated with different clinical characteristics. <strong>a</strong>, An example of the prompt input given to to GPT-4. The phenotypes listed in the second to last sentence (<em>italicised</em>) were changed to allow all HPO phenotypes to be annotated. <strong>b</strong>, Stacked bar plot showing the proportion of the occurrence of each clinical characteristic across all annotated HPO phenotypes. The terms shown on the x-axis are the clinical characteristics for which GPT-4 was asked to determine whether each phenotype caused them.</span>
<span class="glightbox-desc lightbox-desc-2">Figure&nbsp;2: GPT-4 annotations are consistent and accurate across annotations. <strong>a</strong>, Barplot showing the annotation consistency within phenotypes that were annotated more than once. In the lenient metric, annotations were collapsed into two groups (‘always’/‘often’ and ‘never’/‘rarely’). For a given clinical characteristic within a given phenotype, if an annotation was always within the same group it was considered consistent. In the stringent metric, all four annotation categories were considered to be different from one another. Thus, annotations were only defined as consistent if they were all identical. The blue dashed line indicates the probability of two annotations being consistent by chance in the lenient metric (~1/2). The gold dashed line is the probability of two annotations being consistent by chance in the stringent metric (~1/4). <strong>b</strong>, Bar plot of the true positive rate for each annotation. The labels above each bar indicate the number of phenotypes tested.</span>
<span class="glightbox-desc lightbox-desc-3">Figure&nbsp;3: Quantifying the severity of HPO phenotype annotations highlights the most impactful conditions. Heatmap of 10 represetantive phenotypes from each severity class (Profound, Severe, Moderate, Mild) stratified by whether the phenotypes are often/always congenital (<strong>a</strong>-<strong>b</strong>) or rarely/never congenital (<strong>c</strong>-<strong>d</strong>). Continuous severity scores are shown as bars (<strong>b</strong>,<strong>d</strong>) and were calculated by multiplying the numeric values assigned to each clinical characteristic according to <a href="#tbl-metric-weights" class="quarto-xref">Table&nbsp;2</a>. The average normalised score, representing overall phenotype severity on a 0-100 scale, was calculated by aggregating the multiplied values and normalising by the theoretical maximum severity score. The x-axes show each of the clinical characteristics.</span>
<span class="glightbox-desc lightbox-desc-4">Figure&nbsp;4: Distribution of congenital onset across HPO branches. The y-axis shows the proportion of phenotypes that are always/often/rarely/never congenital. The x-axis shows the HPO branch, orderered from highest to lowest proportion of always congenital phenotypes.</span>
<span class="glightbox-desc lightbox-desc-6">Figure&nbsp;5: Relationship between the consistency of GPT-4 annotations (using the stringent criterion) and the level of each phenotype within the HPO ontology (with the number of phenotypes in parentheses). Greater ontology levels (x-axis) indicate more specific phenotypes. The subtitle indicates summary statistics for the overall relationship between HPO level and the proportion of phenotypes that were annotated consistently. The p-values above each bar indicate whether the distribution of consistent/inconsistent annotations, within a given HPO level, significantly deviate from the expected null distribution.</span>
<span class="glightbox-desc lightbox-desc-7">Figure&nbsp;6: Distribution of the composite GPT-4 severity score of the severity scores for all HPO terms.</span>
<span class="glightbox-desc lightbox-desc-8">Figure&nbsp;7: Boxplot showing the relationship between composite severity score (y-axis) and the frequency response categories within each annotation type.</span>
<span class="glightbox-desc lightbox-desc-9">Figure&nbsp;8: Pearson correlations between each individual severity metric and the composite severity score (‘severity_score_gpt’).</span>
<span class="glightbox-desc lightbox-desc-10">Figure&nbsp;9: Distribution of the composite GPT-4 severity score introduced in this paper (y-axis) by an approximation of the severity class system introduced <span class="citation" data-cites="Lazarin2014-rz">Lazarin et al. (<a href="#ref-Lazarin2014-rz" role="doc-biblioref">2014</a>)</span> (x-axis). While these are different schemes for ranking phenotype severity, there is a strong correspondence between them (see summary statistics in subtitle). The sample size (number of phenotypes) is shown in parentheses along the x-axis.</span>
</div>
</section>
</section>
     </main>
<!-- /main column -->  <script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>  </div> <!-- /content -->  <script>var lightboxQuarto = GLightbox({"selector":".lightbox","closeEffect":"zoom","openEffect":"zoom","loop":false,"descPosition":"bottom"});
window.onload = () => {
  lightboxQuarto.on('slide_before_load', (data) => {
    const { slideIndex, slideNode, slideConfig, player, trigger } = data;
    const href = trigger.getAttribute('href');
    if (href !== null) {
      const imgEl = window.document.querySelector(`a[href="${href}"] img`);
      if (imgEl !== null) {
        const srcAttr = imgEl.getAttribute("src");
        if (srcAttr && srcAttr.startsWith("data:")) {
          slideConfig.href = srcAttr;
        }
      }
    } 
  });

  lightboxQuarto.on('slide_after_load', (data) => {
    const { slideIndex, slideNode, slideConfig, player, trigger } = data;
    if (window.Quarto?.typesetMath) {
      window.Quarto.typesetMath(slideNode);
    }
  });

};
          </script> 
  
</body></html>